# 📊 **Lojistik Regresyon ile End-to-End Makine Öğrenmesi Projesi**

**Yönetim Bilişim Sistemleri - Makine Öğrenmesi Ders Notu**

**Tarih:** 12 Şubat 2026  
**Konu:** Lojistik Regresyon ve CRISP-DM Metodolojisi

## 🎯 **Adım 1: Proje Tanıtımı ve CRISP-DM Metodolojisi**

### 📌** Lojistik Regresyon Nedir?**

**Lojistik Regresyon**, <span style='background-color: #FFE5CC; padding: 2px 6px; border-radius: 3px; font-weight: bold;'>sınıflandırma problemleri</span> için kullanılan denetimli bir makine öğrenmesi algoritmasıdır.

#### 🔑 **Temel Özellikler:**

- **İşlevi:** Kategorik sonuçları tahmin etmek
- **Çıktı:** Olasılık değerleri (0 ile 1 arası)
- **Kullanım Alanları:** İkili sınıflandırma (binary classification) veya çoklu sınıflandırma
- **Matematiksel Temel:** Sigmoid (lojistik) fonksiyonu

#### 💼 **YBS Perspektifinden Örnekler:**

| İş Problemi | Hedef Değişken | YBS Uygulaması |
|------------|----------------|----------------|
| 🏦 **Kredi Risk Analizi** | Kredi onaylanacak mı? (Evet/Hayır) | Risk yönetimi sistemleri |
| 🛒 **Müşteri Churn** | Müşteri ayrılacak mı? (Evet/Hayır) | CRM sistemleri |
| 📧 **Spam Tespiti** | Email spam mi? (Spam/Normal) | Email filtreleme sistemleri |
| 🏥 **Hastalık Teşhisi** | Hastalık var mı? (Evet/Hayır) | Sağlık bilgi sistemleri |
| 🎓 **Öğrenci Başarısı** | Sınavı geçecek mi? (Geçer/Kalır) | Eğitim yönetim sistemleri |

### 📐 **Lojistik Regresyon vs Lineer Regresyon**

```mermaid
graph LR
    A[Regresyon Modelleri] --> B[Lineer Regresyon]
    A --> C[Lojistik Regresyon]
    
    B --> D[Sürekli Değerler<br/>Tahmin Eder]
    B --> E[Örnek: Ev Fiyatı<br/>Maaş Tahmini]
    
    C --> F[Kategorik Değerler<br/>Sınıflandırır]
    C --> G[Örnek: Evet/Hayır<br/>Başarılı/Başarısız]
    
    style A fill:#E8F4F8,stroke:#2C5F7C,stroke-width:2px,color:#000
    style B fill:#FFF4E6,stroke:#D68910,stroke-width:2px,color:#000
    style C fill:#E6F4EA,stroke:#34A853,stroke-width:2px,color:#000
    style D fill:#FFF9E6,stroke:#F4B400,stroke-width:1px,color:#000
    style E fill:#FFF9E6,stroke:#F4B400,stroke-width:1px,color:#000
    style F fill:#F0FFF4,stroke:#5CB85C,stroke-width:1px,color:#000
    style G fill:#F0FFF4,stroke:#5CB85C,stroke-width:1px,color:#000
```

### 🔬 **CRISP-DM Metodolojisi**

**CRISP-DM** (Cross-Industry Standard Process for Data Mining), veri madenciliği ve makine öğrenmesi projelerinde kullanılan <span style='background-color: #E6F4EA; padding: 2px 6px; border-radius: 3px; font-weight: bold;'>endüstri standardı bir süreç modelidir</span>.

#### 📊 **CRISP-DM'nin 6 Aşaması:**

```mermaid
graph TB
    A[1. Business Understanding<br/>İş Anlayışı] --> B[2. Data Understanding<br/>Veri Anlayışı]
    B --> C[3. Data Preparation<br/>Veri Hazırlama]
    C --> D[4. Modeling<br/>Modelleme]
    D --> E[5. Evaluation<br/>Değerlendirme]
    E --> F[6. Deployment<br/>Canlıya Alma]
    
    E -.Geri Dönüş.-> A
    E -.Geri Dönüş.-> B
    E -.Geri Dönüş.-> C
    E -.Geri Dönüş.-> D
    
    style A fill:#FFE5CC,stroke:#D68910,stroke-width:3px,color:#000
    style B fill:#E6F4EA,stroke:#34A853,stroke-width:3px,color:#000
    style C fill:#E8F4F8,stroke:#2C5F7C,stroke-width:3px,color:#000
    style D fill:#F4E6FF,stroke:#8E44AD,stroke-width:3px,color:#000
    style E fill:#FFE6E6,stroke:#C0392B,stroke-width:3px,color:#000
    style F fill:#FFF9E6,stroke:#F39C12,stroke-width:3px,color:#000
```

### 📋 **CRISP-DM Aşamalarının Detaylı Açıklaması**

#### <span style='background-color: #FFE5CC; padding: 4px 8px; border-radius: 4px;'>1️⃣ Business Understanding (İş Anlayışı)</span>

**Amaç:** Projenin iş hedeflerini anlamak ve makine öğrenmesi problemine dönüştürmek.

**Yapılacaklar:**
- 🎯 İş probleminin net tanımı
- 📈 Başarı kriterlerinin belirlenmesi
- 💰 Maliyet-fayda analizi
- 📊 KPI (Key Performance Indicator) tanımları

**YBS Örneği:** Bir e-ticaret şirketi müşteri kaybını azaltmak istiyor → Churn tahmini modeli geliştirilmeli.

#### <span style='background-color: #E6F4EA; padding: 4px 8px; border-radius: 4px;'>2️⃣ Data Understanding (Veri Anlayışı)</span>

**Amaç:** Veriyi toplamak, keşfetmek ve kalite problemlerini tespit etmek.

**Yapılacaklar:**
- 📥 Veri toplama ve yükleme
- 🔍 Keşifsel Veri Analizi (EDA)
- 📊 İstatistiksel özetler
- 🎨 Görselleştirmeler
- ❓ Veri kalitesi değerlendirmesi

#### <span style='background-color: #E8F4F8; padding: 4px 8px; border-radius: 4px;'>3️⃣ Data Preparation (Veri Hazırlama)</span>

**Amaç:** Veriyi modelleme için hazır hale getirmek.

**Yapılacaklar:**
- 🧹 Veri temizleme (eksik değerler, aykırı değerler)
- 🔧 Feature Engineering (yeni öznitelik türetme)
- 🔄 Dönüşümler (encoding, scaling)
- ✂️ Feature Selection (özellik seçimi)
- 📦 Train-Test split

**Not:** Bu aşama genellikle en çok zaman alan aşamadır (%60-80).

#### <span style='background-color: #F4E6FF; padding: 4px 8px; border-radius: 4px;'>4️⃣ Modeling (Modelleme)</span>

**Amaç:** Makine öğrenmesi modelini seçmek, eğitmek ve optimize etmek.

**Yapılacaklar:**
- 🤖 Model seçimi (Lojistik Regresyon, KNN, SVM, vb.)
- 🎓 Model eğitimi
- 🔧 Hiperparametre optimizasyonu
- ✅ Cross-validation
- 📊 Farklı modellerin karşılaştırılması

#### <span style='background-color: #FFE6E6; padding: 4px 8px; border-radius: 4px;'>5️⃣ Evaluation (Değerlendirme)</span>

**Amaç:** Modelin performansını değerlendirmek ve iş hedeflerini karşılayıp karşılamadığını kontrol etmek.

**Yapılacaklar:**
- 📈 Performans metrikleri (Accuracy, Precision, Recall, F1-Score, ROC-AUC)
- 🔍 Confusion Matrix analizi
- 💼 İş değeri analizi
- ✔️ Model validasyonu
- 🔄 Gerekirse önceki aşamalara geri dönüş

#### <span style='background-color: #FFF9E6; padding: 4px 8px; border-radius: 4px;'>6️⃣ Deployment (Canlıya Alma)</span>

**Amaç:** Modeli üretim ortamına almak ve kullanıma sunmak.

**Yapılacaklar:**
- 💾 Model kaydetme (pickle, joblib)
- 🚀 Model deployment (API, web servisi)
- 📡 Monitoring sistemi kurma
- 📝 Dokümantasyon
- 🔄 Model güncelleme stratejisi

**Not:** Bu projede Deployment aşamasına kadar ilerleyeceğiz, model kaydedilerek hazır hale getirilecek.

### 🎓 **YBS Öğrencileri İçin Önemli Notlar**

> **💡 Not 1:** CRISP-DM doğrusal bir süreç değildir. İteratif (yinelemeli) bir yaklaşımdır. Değerlendirme aşamasında yetersiz sonuçlar alırsanız, önceki aşamalara geri dönebilirsiniz.

> **💡 Not 2:** İş dünyasında makine öğrenmesi projeleri her zaman teknik bir problem değil, öncelikle bir **iş problemi**dir. Bu yüzden Business Understanding aşaması kritik öneme sahiptir.

> **💡 Not 3:** Veri bilimciler zamanlarının çoğunu (%60-80) Data Preparation aşamasında harcarlar. Temiz veri = Başarılı model!

## 📦 **Adım 2: Kaggle Dataset Seçimi ve Veriyi Lokale İndirme**

Bu adımda, **Kaggle**'dan lojistik regresyon için uygun popüler bir veri seti seçeceğiz ve lokale indireceğiz.

### 🎲 **Seçilen Dataset: Bank Marketing Campaign**

**Dataset Adı:** Bank Marketing Dataset  
**Kaynak:** UCI Machine Learning Repository (Kaggle üzerinden)  
**Veri Sayısı:** ~45,000+ kayıt  
**Problem Tipi:** İkili Sınıflandırma (Binary Classification)

#### 📊 **İş Problemi**

Bir **Portekiz bankası**, müşterilerine telefon ile **vadeli mevduat (term deposit)** ürünü pazarlamak için kampanya düzenliyor. 

**YBS Perspektifi:**
- 📞 **Call Center Yönetimi:** Hangi müşterilere öncelik verilmeli?
- 💰 **Maliyetleri Azaltma:** Başarısız aramalara zaman harcanmasın
- 🎯 **Hedef Kitle Belirleme:** Doğru müşteriyi doğru zamanda ara
- 📈 **Kampanya ROI'sini Artırma:** Dönüşüm oranını maksimize et

#### 🎯 **Tahmin Hedefi**

**Hedef Değişken (y):** Müşteri vadeli mevduat açacak mı?
- ✅ **Yes (Evet)** = Müşteri ürünü satın aldı
- ❌ **No (Hayır)** = Müşteri ürünü satın almadı

```mermaid
graph LR
    A[Müşteri Bilgileri] --> B[Lojistik Regresyon<br/>Modeli]
    B --> C{Tahmin}
    C --> D[Yes<br/>Ürünü Alacak]
    C --> E[No<br/>Ürünü Almayacak]
    
    D --> F[Call Center'a<br/>Öncelikli Müşteri]
    E --> G[Kampanyaya<br/>Dahil Etme]
    
    style A fill:#E8F4F8,stroke:#2C5F7C,stroke-width:2px,color:#000
    style B fill:#F4E6FF,stroke:#8E44AD,stroke-width:3px,color:#000
    style C fill:#FFF9E6,stroke:#F39C12,stroke-width:2px,color:#000
    style D fill:#E6F4EA,stroke:#34A853,stroke-width:2px,color:#000
    style E fill:#FFE6E6,stroke:#C0392B,stroke-width:2px,color:#000
    style F fill:#F0FFF4,stroke:#5CB85C,stroke-width:1px,color:#000
    style G fill:#FFF5F5,stroke:#E74C3C,stroke-width:1px,color:#000
```

### 📋 **Dataset Özellikleri**

Veri seti **20 bağımsız değişken** ve **1 hedef değişken** içermektedir:

#### 👤 **Müşteri Demografik Bilgileri**
- **age**: Yaş
- **job**: Meslek (admin, technician, services, vb.)
- **marital**: Medeni durum (married, single, divorced)
- **education**: Eğitim seviyesi

#### 💰 **Finansal Bilgiler**
- **default**: Kredi temerrüdü var mı?
- **balance**: Yıllık ortalama bakiye
- **housing**: Konut kredisi var mı?
- **loan**: Bireysel kredi var mı?

#### 📞 **Kampanya İletişim Bilgileri**
- **contact**: İletişim tipi (cellular, telephone)
- **day**: Son iletişim günü
- **month**: Son iletişim ayı
- **duration**: Son görüşme süresi (saniye)
- **campaign**: Bu kampanyada yapılan iletişim sayısı
- **pdays**: Önceki kampanyadan sonra geçen gün sayısı
- **previous**: Önceki kampanyalarda yapılan iletişim sayısı
- **poutcome**: Önceki kampanya sonucu

#### 📊 **Ekonomik Göstergeler**
- **emp.var.rate**: İstihdam değişim oranı
- **cons.price.idx**: Tüketici fiyat endeksi
- **cons.conf.idx**: Tüketici güven endeksi
- **euribor3m**: 3 aylık euribor oranı
- **nr.employed**: Çalışan sayısı

### 📥 **Gerekli Kütüphaneleri İçe Aktarma**

Öncelikle projemizde kullanacağımız tüm Python kütüphanelerini içe aktaralım:

In [1]:
# Veri işleme ve analiz kütüphaneleri
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

# Görselleştirme kütüphaneleri
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Scikit-learn - Veri hazırlama
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV, StratifiedKFold
from sklearn.preprocessing import StandardScaler, LabelEncoder, MinMaxScaler

# Scikit-learn - Makine öğrenmesi
from sklearn.linear_model import LogisticRegression

# Scikit-learn - Performans metrikleri
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score,
    confusion_matrix, classification_report, roc_auc_score, roc_curve
)

# Model kaydetme
import pickle
import joblib

# Dosya işlemleri
import os
from datetime import datetime

# Grafik ayarları
plt.style.use('seaborn-v0_8-pastel')
sns.set_palette("pastel")

print("✅ Tüm kütüphaneler başarıyla yüklendi!")
print(f"📅 Tarih: {datetime.now().strftime('%d-%m-%Y %H:%M:%S')}")

✅ Tüm kütüphaneler başarıyla yüklendi!
📅 Tarih: 12-02-2026 00:40:25


### 🌐 **Veri Setini İndirme ve Kaydetme**

Veri setini URL üzerinden pandas ile yükleyip, lokale kaydedeceğiz:

In [12]:
# Veri setini yükleme - Akıllı separator algılama ile
# Not: Farklı kaynaklardaki veri setleri farklı ayırıcılar kullanabilir (virgül veya noktalı virgül)

data_folder = "data"
if not os.path.exists(data_folder):
    os.makedirs(data_folder)
    print(f"📁 '{data_folder}' klasörü oluşturuldu")

local_file_path = os.path.join(data_folder, "bank_marketing.csv")

# Veriyi yüklemek için yardımcı fonksiyon - doğru separator'ı otomatik algılar
def load_bank_data(file_path_or_url, is_url=False):
    """Veriyi yükler ve doğru separator'ı otomatik algılar"""
    separators = [';', ',']  # Olası ayırıcılar
    
    for sep in separators:
        try:
            if is_url:
                df_test = pd.read_csv(file_path_or_url, sep=sep, nrows=5)
            else:
                df_test = pd.read_csv(file_path_or_url, sep=sep, nrows=5)
            
            # Eğer çok fazla sütun varsa (>15), doğru separator bulunmuştur
            if df_test.shape[1] > 15:
                print(f"✅ Doğru ayırıcı bulundu: '{sep}'")
                if is_url:
                    return pd.read_csv(file_path_or_url, sep=sep)
                else:
                    return pd.read_csv(file_path_or_url, sep=sep)
        except:
            continue
    
    # Hiçbir separator çalışmazsa, varsayılan olarak oku
    print("⚠️ Separator otomatik algılanamadı, varsayılan ayarlarla yükleniyor...")
    return pd.read_csv(file_path_or_url)

# Eğer daha önce indirilmediyse, alternatif kaynaktan indir
if not os.path.exists(local_file_path):
    print("📥 Veri seti indiriliyor...")
    
    # Alternatif URL'ler - sırayla denenecek
    urls = [
        "https://raw.githubusercontent.com/madmashup/targeted-marketing-predictive-engine/master/banking.csv",
        "https://raw.githubusercontent.com/stedy/Machine-Learning-with-R-datasets/master/bank.csv"
    ]
    
    df = None
    for i, url in enumerate(urls):
        try:
            print(f"🔗 Kaynak {i+1} deneniyor...")
            df = load_bank_data(url, is_url=True)
            
            # Veri setinin geçerli olup olmadığını kontrol et
            if df is not None and df.shape[1] > 10:
                print(f"✅ Veri seti başarıyla yüklendi! ({df.shape[0]} satır, {df.shape[1]} sütun)")
                break
        except Exception as e:
            print(f"❌ Kaynak {i+1} başarısız: {str(e)[:50]}...")
            continue
    
    # Eğer hiçbir URL çalışmazsa
    if df is None or df.shape[1] <= 10:
        print("\n" + "="*70)
        print("⚠️ OTOMATİK İNDİRME BAŞARISIZ!")
        print("="*70)
        print("\n📌 Lütfen veri setini manuel olarak indirin:")
        print("   1. https://www.kaggle.com/datasets/janiobachmann/bank-marketing-dataset")
        print("   2. 'bank.csv' dosyasını indirin")
        print(f"   3. '{data_folder}' klasörüne 'bank_marketing.csv' adıyla kaydedin")
        print("="*70)
        raise FileNotFoundError("Veri seti indirilemedi. Lütfen manuel olarak indirin.")
    
    # CSV olarak kaydetme
    df.to_csv(local_file_path, index=False)
    print(f"💾 Veri seti lokale kaydedildi: {local_file_path}")
    
else:
    # Daha önce indirilmiş veriyi yükle
    print("📂 Lokalde kaydedilmiş veri seti bulundu, yükleniyor...")
    df = load_bank_data(local_file_path, is_url=False)
    
    # Eğer yanlış parse edilmişse (tek sütun varsa), dosyayı sil ve yeniden indir
    if df.shape[1] <= 2:
        print("⚠️ Veri seti yanlış formatlanmış, yeniden indiriliyor...")
        os.remove(local_file_path)
        
        # Yeniden indir
        urls = [
            "https://raw.githubusercontent.com/madmashup/targeted-marketing-predictive-engine/master/banking.csv",
            "https://raw.githubusercontent.com/stedy/Machine-Learning-with-R-datasets/master/bank.csv"
        ]
        
        for i, url in enumerate(urls):
            try:
                print(f"🔗 Kaynak {i+1} deneniyor...")
                df = load_bank_data(url, is_url=True)
                if df.shape[1] > 10:
                    print(f"✅ Veri seti başarıyla yüklendi!")
                    df.to_csv(local_file_path, index=False)
                    print(f"💾 Veri seti lokale kaydedildi: {local_file_path}")
                    break
            except Exception as e:
                continue
    else:
        print("✅ Veri seti başarıyla yüklendi!")

# Veri seti hakkında temel bilgiler
print("\n" + "="*60)
print("📊 VERİ SETİ TEMEL BİLGİLERİ")
print("="*60)
print(f"📏 Satır Sayısı    : {df.shape[0]:,}")
print(f"📏 Sütun Sayısı    : {df.shape[1]}")
print(f"💾 Bellek Kullanımı: {df.memory_usage(deep=True).sum() / 1024**2:.2f} MB")
print(f"🎯 Hedef Değişken  : 'y' (Vadeli mevduat açacak mı?)")
print("="*60)

📂 Lokalde kaydedilmiş veri seti bulundu, yükleniyor...
✅ Doğru ayırıcı bulundu: ','
✅ Veri seti başarıyla yüklendi!

📊 VERİ SETİ TEMEL BİLGİLERİ
📏 Satır Sayısı    : 41,188
📏 Sütun Sayısı    : 21
💾 Bellek Kullanımı: 25.11 MB
🎯 Hedef Değişken  : 'y' (Vadeli mevduat açacak mı?)


### 👀 **Veri Setinin İlk Görünümü**

Veri setinin ilk birkaç satırına göz atalım:

In [13]:
# İlk 5 satırı görüntüleme
print("🔍 Veri Setinin İlk 5 Satırı:\n")
df.head()

🔍 Veri Setinin İlk 5 Satırı:



,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp_var_rate,cons_price_idx,cons_conf_idx,euribor3m,nr_employed,y
0,44,blue-collar,married,basic.4y,unknown,yes,no,cellular,aug,thu,...,1,999,0,nonexistent,1.4,93.444,-36.1,4.963,5228.1,0
1,53,technician,married,unknown,no,no,no,cellular,nov,fri,...,1,999,0,nonexistent,-0.1,93.200,-42.0,4.021,5195.8,0
2,28,management,single,university.degree,no,yes,no,cellular,jun,thu,...,3,6,2,success,-1.7,94.055,-39.8,0.729,4991.6,1
3,39,services,married,high.school,no,no,no,cellular,apr,fri,...,2,999,0,nonexistent,-1.8,93.075,-47.1,1.405,5099.1,0
4,55,retired,married,basic.4y,no,yes,no,cellular,aug,fri,...,1,3,1,success,-2.9,92.201,-31.4,0.869,5076.2,1


In [14]:
# Sütun bilgilerini görüntüleme
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp_var_rate    41188 non-null  float64
 16  cons_price_idx  41188 non-null  float64
 17  cons_conf_idx   41188 non-null 

### 🎯 **Hedef Değişken Dağılımı**

Tahmin edeceğimiz hedef değişkenin (**y**) dağılımına bakalım:

In [15]:
# Hedef değişkenin benzersiz değerlerini kontrol edelim
print("🔍 Hedef değişken (y) içindeki benzersiz değerler:")
print(df['y'].unique())
print(f"\nVeri tipi: {df['y'].dtype}")
print(f"\nDağılım:")
print(df['y'].value_counts())

🔍 Hedef değişken (y) içindeki benzersiz değerler:
[0 1]

Veri tipi: int64

Dağılım:
y
0    36548
1     4640
Name: count, dtype: int64


In [16]:
# Hedef değişkenin dağılımı
target_counts = df['y'].value_counts()
target_percentages = df['y'].value_counts(normalize=True) * 100

# Hedef değişkenin değerlerini otomatik algılama
target_values = target_counts.index.tolist()
# Genellikle 'no', 0, 'n' gibi değerler negatif, 'yes', 1 gibi değerler pozitif
negative_value = target_values[0]  # İlk değer (genellikle çoğunluk sınıfı)
positive_value = target_values[1] if len(target_values) > 1 else target_values[0]

print("📊 HEDEF DEĞİŞKEN (y) DAĞILIMI")
print("="*60)
print(f"❌ {negative_value} (Ürünü Almayan)  : {target_counts.iloc[0]:,} ({target_percentages.iloc[0]:.2f}%)")
print(f"✅ {positive_value} (Ürünü Alan)      : {target_counts.iloc[1]:,} ({target_percentages.iloc[1]:.2f}%)")
print("="*60)

# Görselleştirme
fig = make_subplots(
    rows=1, cols=2,
    subplot_titles=('Hedef Degisken Dagilimi', 'Yuzdelik Oran'),
    specs=[[{"type": "bar"}, {"type": "pie"}]]
)

# Bar chart
fig.add_trace(
    go.Bar(
        x=[f'{negative_value} (Almayan)', f'{positive_value} (Alan)'],
        y=[target_counts.iloc[0], target_counts.iloc[1]],
        marker_color=['#FFB3BA', '#BAFFC9'],
        text=[f"{target_counts.iloc[0]:,}", f"{target_counts.iloc[1]:,}"],
        textposition='outside',
        showlegend=False
    ),
    row=1, col=1
)

# Pie chart
fig.add_trace(
    go.Pie(
        labels=[f'{negative_value} (Almayan)', f'{positive_value} (Alan)'],
        values=[target_counts.iloc[0], target_counts.iloc[1]],
        marker_colors=['#FFB3BA', '#BAFFC9'],
        textinfo='label+percent',
        textfont_size=12
    ),
    row=1, col=2
)

fig.update_layout(
    title_text="Vadeli Mevduat Urun Satin Alma Durumu",
    title_font_size=16,
    height=400,
    showlegend=False
)

fig.show()

# YBS perspektifinden değerlendirme
print("\n💼 YBS PERSPEKTİFİNDEN DEĞERLENDİRME:")
print("="*60)
print(f"⚠️ Veri dengesizlik var! (Class Imbalance)")
print(f"   - Çoğunluk sınıfı: {negative_value} ({target_percentages.iloc[0]:.1f}%)")
print(f"   - Azınlık sınıfı : {positive_value} ({target_percentages.iloc[1]:.1f}%)")
print(f"\n📌 İş Etkisi:")
print(f"   Müşterilerin sadece %{target_percentages.iloc[1]:.1f}'i ürünü satın alıyor.")
print(f"   Bu da call center operasyonlarının verimliliğinin düşük olduğunu gösteriyor.")
print("="*60)

📊 HEDEF DEĞİŞKEN (y) DAĞILIMI
❌ 0 (Ürünü Almayan)  : 36,548 (88.73%)
✅ 1 (Ürünü Alan)      : 4,640 (11.27%)



💼 YBS PERSPEKTİFİNDEN DEĞERLENDİRME:
⚠️ Veri dengesizlik var! (Class Imbalance)
   - Çoğunluk sınıfı: 0 (88.7%)
   - Azınlık sınıfı : 1 (11.3%)

📌 İş Etkisi:
   Müşterilerin sadece %11.3'i ürünü satın alıyor.
   Bu da call center operasyonlarının verimliliğinin düşük olduğunu gösteriyor.


## 💼 **Adım 3: Business Understanding (İş Anlayışı)**

**CRISP-DM'nin ilk ve en kritik aşaması:** İş problemini anlamak ve makine öğrenmesi problemine dönüştürmek.

> **💡 YBS Notu:** Bu aşama, teknik olmaktan çok iş odaklıdır. Bir YBS uzmanı olarak, teknolojiyi iş ihtiyaçlarıyla birleştirme yeteneğiniz burada öne çıkar.

### 🎯 **İş Problemi Tanımı**

**Senaryo:** Bir **Portekiz bankası**, müşterilerine telefon pazarlama kampanyası ile **vadeli mevduat (term deposit)** ürünü satmaya çalışıyor.

#### 📌 **Mevcut Durum ve Sorunlar:**

**Problem 1: Düşük Dönüşüm Oranı**
- Müşterilerin sadece %11,7'si vadeli mevduat açıyor
- Yüksek miktarda başarısız arama yapılıyor
- Call center kaynakları verimsiz kullanılıyor

**Problem 2: Yüksek Operasyonel Maliyetler**
- Her arama için zaman ve insan kaynağı harcanıyor
- İlgilenmeyecek müşterilere gereksiz aramalar yapılıyor
- ROI (Return on Investment) düşük

**Problem 3: Müşteri Deneyimi**
- İlgilenmeyen müşterilere yapılan aramalar memnuniyetsizlik yaratıyor
- Müşteri kaybı riski var

```mermaid
graph TB
    A[Mevcut Durum<br/>Tüm Müşterilere Arama] --> B{Sonuç}
    B --> C[%11.7 Başarılı<br/>Ürün Satın Alıyor]
    B --> D[%88.3 Başarısız<br/>Ürün Almıyor]
    
    D --> E[Kaynak İsrafı<br/>Yüksek Maliyet<br/>Düşük ROI]
    
    style A fill:#FFE5CC,stroke:#D68910,stroke-width:2px,color:#000
    style B fill:#FFF9E6,stroke:#F39C12,stroke-width:2px,color:#000
    style C fill:#E6F4EA,stroke:#34A853,stroke-width:2px,color:#000
    style D fill:#FFE6E6,stroke:#C0392B,stroke-width:2px,color:#000
    style E fill:#FFB3BA,stroke:#E74C3C,stroke-width:2px,color:#000
```

### 💡 **Çözüm Önerisi: Tahmine Dayalı Model**

**Makine Öğrenmesi ile Çözüm:**

Geçmiş kampanya verilerini kullanarak, **hangi müşterilerin vadeli mevduat açma olasılığının yüksek olduğunu tahmin eden** bir model geliştirmek.

```mermaid
graph TB
    A[Hedef Sistem<br/>Akıllı Müşteri Seçimi] --> B[ML Modeli<br/>Tahmin Yapar]
    B --> C{Olasılık}
    C --> D[Yüksek Olasılık<br/>%70+]
    C --> E[Düşük Olasılık<br/>%30-]
    
    D --> F[Call Center'a Gönder<br/>Öncelikli Arama]
    E --> G[Aramayı Ertele<br/>veya Farklı Kanal Kullan]
    
    F --> H[Artan Dönüşüm<br/>Azalan Maliyet<br/>Yüksek ROI]
    
    style A fill:#E6F4EA,stroke:#34A853,stroke-width:3px,color:#000
    style B fill:#F4E6FF,stroke:#8E44AD,stroke-width:3px,color:#000
    style C fill:#FFF9E6,stroke:#F39C12,stroke-width:2px,color:#000
    style D fill:#BAFFC9,stroke:#28A745,stroke-width:2px,color:#000
    style E fill:#FFE6E6,stroke:#C0392B,stroke-width:2px,color:#000
    style F fill:#D4EDDA,stroke:#155724,stroke-width:2px,color:#000
    style G fill:#FFF3CD,stroke:#856404,stroke-width:2px,color:#000
    style H fill:#CCE5FF,stroke:#004085,stroke-width:3px,color:#000
```

#### 🎯 **Proje Hedefleri**

| Hedef | Açıklama | Ölçülebilir Metrik |
|-------|----------|-------------------|
| **1. Dönüşüm Oranını Artırmak** | Ürün satın alan müşteri oranını yükseltmek | Dönüşüm oranı: %11.7 → %20+ |
| **2. Maliyetleri Azaltmak** | Gereksiz aramaları minimize etmek | Arama sayısı: 50% azaltma |
| **3. ROI'yi İyileştirmek** | Kampanya getirisini maksimize etmek | ROI: 2x artış hedefi |
| **4. Müşteri Deneyimini İyileştirmek** | İlgisiz müşterilere daha az arama | Müşteri şikayeti: %30 azalma |
| **5. Karar Destek Sistemi** | Veri odaklı karar verme altyapısı | Model accuracy: %85+ |

### 👥 **Stakeholder (Paydaş) Analizi**

**YBS Perspektifi:** Başarılı bir proje için tüm paydaşların ihtiyaçlarını anlamalıyız.

```mermaid
graph TD
    A[Stakeholder'lar] --> B[Üst Yönetim<br/>CEO, CFO]
    A --> C[Pazarlama Müdürü]
    A --> D[Call Center Müdürü]
    A --> E[IT Direktörü]
    A --> F[Müşteriler]
    
    B --> B1[ROI Artışı<br/>Stratejik Hedefler]
    C --> C1[Kampanya Başarısı<br/>Dönüşüm Oranları]
    D --> D1[Operasyonel Verimlilik<br/>Çalışan Performansı]
    E --> E1[Teknik Altyapı<br/>Sistem Entegrasyonu]
    F --> F1[Daha İyi Deneyim<br/>İlgili Teklifler]
    
    style A fill:#FFE5CC,stroke:#D68910,stroke-width:3px,color:#000
    style B fill:#E8F4F8,stroke:#2C5F7C,stroke-width:2px,color:#000
    style C fill:#E6F4EA,stroke:#34A853,stroke-width:2px,color:#000
    style D fill:#F4E6FF,stroke:#8E44AD,stroke-width:2px,color:#000
    style E fill:#FFF9E6,stroke:#F39C12,stroke-width:2px,color:#000
    style F fill:#FFE6E6,stroke:#C0392B,stroke-width:2px,color:#000
    style B1 fill:#F0F8FF,stroke:#4A90E2,stroke-width:1px,color:#000
    style C1 fill:#F0FFF4,stroke:#5CB85C,stroke-width:1px,color:#000
    style D1 fill:#FFF0F5,stroke:#9B59B6,stroke-width:1px,color:#000
    style E1 fill:#FFF9E6,stroke:#F4B400,stroke-width:1px,color:#000
    style F1 fill:#FFF5F5,stroke:#E74C3C,stroke-width:1px,color:#000
```

#### 📋 **Paydaş İhtiyaçları**

| Paydaş | İhtiyaç | Beklenti |
|--------|---------|----------|
| **👔 Üst Yönetim** | Karlılık artışı, stratejik değer | ROI'de ölçülebilir iyileşme |
| **📊 Pazarlama Müdürü** | Kampanya başarısı, hedef kitle optimizasyonu | Dönüşüm oranında %50+ artış |
| **📞 Call Center Müdürü** | Verimli kaynak kullanımı, çalışan motivasyonu | Başarılı arama oranında artış |
| **💻 IT Direktörü** | Entegrasyon kolaylığı, sistem performansı | Sorunsuz sistem entegrasyonu |
| **👥 Müşteriler** | İlgili teklifler, rahatsız edilmeme | Daha az gereksiz arama |

### 📊 Başarı Kriterleri ve KPI'lar

**KPI (Key Performance Indicator):** Proje başarısını ölçmek için kullanılacak temel göstergeler.

#### 🎯 Teknik Başarı Kriterleri

| Metrik | Açıklama | Hedef Değer | Öncelik |
|--------|----------|-------------|---------|
| **Accuracy (Doğruluk)** | Genel tahmin başarısı | ≥ %85 | 🔴 Yüksek |
| **Precision (Kesinlik)** | Pozitif tahminlerin doğruluğu | ≥ %75 | 🟡 Orta |
| **Recall (Duyarlılık)** | Gerçek pozitifleri yakalama | ≥ %70 | 🔴 Yüksek |
| **F1-Score** | Precision ve Recall dengesi | ≥ %72 | 🔴 Yüksek |
| **AUC-ROC** | Model ayırt etme gücü | ≥ 0.85 | 🟡 Orta |

**Not:** <span style='background-color: #FFE5CC; padding: 2px 6px; border-radius: 3px;'>Recall yüksek olmalı!</span> Çünkü potansiyel müşteriyi kaçırmak, yanlış tahmin etmekten daha maliyetli.

#### 💼 **İş Başarı Kriterleri**

| KPI | Mevcut Durum | Hedef | Ölçüm Yöntemi |
|-----|--------------|-------|---------------|
| **Dönüşüm Oranı** | %11.7 | %18-20 | (Satış / Arama Sayısı) × 100 |
| **Görüşme Başına Maliyet** | €10 | €6-7 | Toplam Maliyet / Görüşme Sayısı |
| **ROI (Yatırım Getirisi)** | 1.5x | 3.0x | (Gelir - Maliyet) / Maliyet |
| **Müşteri Memnuniyeti** | 6.5/10 | 8.0/10 | Anket skorları |
| **Operasyonel Verimlilik** | %40 | %65 | Başarılı Arama / Toplam Arama |



```mermaid
graph LR
    A[Model Metrikleri] --> B[İş Metrikleri]
    B --> C[Finansal Sonuçlar]
    
    A1[Accuracy: 85%+] --> B1[Dönüşüm: 20%+]
    A2[Recall: 70%+] --> B2[Kayıp Önleme]
    A3[Precision: 75%+] --> B3[Verimlilik]
    
    B1 --> C1[Gelir Artışı]
    B2 --> C2[Fırsat Maliyeti Azalması]
    B3 --> C3[Operasyonel Tasarruf]
    
    C1 --> D[ROI: 3.0x]
    C2 --> D
    C3 --> D
    
    style A fill:#F4E6FF,stroke:#8E44AD,stroke-width:2px,color:#000
    style B fill:#E6F4EA,stroke:#34A853,stroke-width:2px,color:#000
    style C fill:#FFE5CC,stroke:#D68910,stroke-width:2px,color:#000
    style D fill:#CCE5FF,stroke:#004085,stroke-width:3px,color:#000
    style A1 fill:#FFF0F5,stroke:#9B59B6,stroke-width:1px,color:#000
    style A2 fill:#FFF0F5,stroke:#9B59B6,stroke-width:1px,color:#000
    style A3 fill:#FFF0F5,stroke:#9B59B6,stroke-width:1px,color:#000
    style B1 fill:#F0FFF4,stroke:#5CB85C,stroke-width:1px,color:#000
    style B2 fill:#F0FFF4,stroke:#5CB85C,stroke-width:1px,color:#000
    style B3 fill:#F0FFF4,stroke:#5CB85C,stroke-width:1px,color:#000
    style C1 fill:#FFF9E6,stroke:#F4B400,stroke-width:1px,color:#000
    style C2 fill:#FFF9E6,stroke:#F4B400,stroke-width:1px,color:#000
    style C3 fill:#FFF9E6,stroke:#F4B400,stroke-width:1px,color:#000
```

### 💰 **Maliyet-Fayda Analizi**

**YBS Perspektifi:** Her proje için ROI hesaplaması kritiktir.

#### 📉 Proje Maliyetleri (Tahmini)

| Maliyet Kalemi | Tutar (€) | Açıklama |
|----------------|-----------|----------|
| Veri Bilimci (3 ay) | 18,000 | Analiz, modelleme, deployment |
| IT Altyapısı | 5,000 | Sunucu, yazılım lisansları |
| Entegrasyon Maliyeti | 8,000 | CRM ve call center sistemi entegrasyonu |
| Test ve Eğitim | 4,000 | Kullanıcı eğitimi, pilot çalışma |
| **TOPLAM** | **35,000** | İlk yıl toplam yatırım |

#### 📈 Beklenen Faydalar (Yıllık)

| Fayda Kalemi | Tutar (€) | Hesaplama |
|--------------|-----------|-----------|
| Artan Dönüşümlerden Gelir | 85,000 | Ek 500 müşteri × €170 ortalama kazanç |
| Operasyonel Maliyet Tasarrufu | 45,000 | 50% daha az arama × €10 arama maliyeti |
| Müşteri Memnuniyeti Değeri | 15,000 | Azalan şikayet, artan sadakat |
| **TOPLAM FAYDA** | **145,000** | Yıllık toplam fayda |

#### 🎯 **ROI Hesaplaması**

In [21]:
# Maliyet-Fayda Analizi Hesaplaması

# Maliyetler
toplam_maliyet = 35000  # Euro
veri_bilimci_maliyeti = 18000
it_altyapi = 5000
entegrasyon = 8000
test_egitim = 4000

# Faydalar (Yıllık)
artan_donusum_geliri = 85000  # Euro
operasyonel_tasarruf = 45000
musteri_memnuniyeti_degeri = 15000
toplam_fayda = artan_donusum_geliri + operasyonel_tasarruf + musteri_memnuniyeti_degeri

# ROI Hesaplama
net_fayda = toplam_fayda - toplam_maliyet
roi = (net_fayda / toplam_maliyet) * 100
geri_odeme_suresi = toplam_maliyet / (toplam_fayda / 12)  # Ay cinsinden

print("💰 MALİYET-FAYDA ANALİZİ SONUÇLARI")
print("="*60)
print(f"📉 Toplam Proje Maliyeti    : €{toplam_maliyet:,}")
print(f"📈 Yıllık Toplam Fayda      : €{toplam_fayda:,}")
print(f"💵 Net Fayda (İlk Yıl)      : €{net_fayda:,}")
print(f"📊 ROI (Yatırım Getirisi)   : %{roi:.1f}")
print(f"⏱️  Geri Ödeme Süresi        : {geri_odeme_suresi:.1f} ay (~{geri_odeme_suresi/12:.2f} yıl)")
print("="*60)

# Görselleştirme
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=(
        'Maliyet Dagilimi', 
        'Fayda Dagilimi',
        'Maliyet vs Fayda Karsilastirmasi',
        'ROI Gorsellestirilmesi'
    ),
    specs=[
        [{"type": "pie"}, {"type": "pie"}],
        [{"type": "bar"}, {"type": "indicator"}]
    ]
)

# 1. Maliyet Dağılımı (Pie Chart)
fig.add_trace(
    go.Pie(
        labels=['Veri Bilimci', 'IT Altyapı', 'Entegrasyon', 'Test & Eğitim'],
        values=[veri_bilimci_maliyeti, it_altyapi, entegrasyon, test_egitim],
        marker_colors=['#FFB3BA', '#FFDFBA', '#FFFFBA', '#BAFFC9'],
        hole=0.3
    ),
    row=1, col=1
)

# 2. Fayda Dağılımı (Pie Chart)
fig.add_trace(
    go.Pie(
        labels=['Artan Dönüşüm', 'Operasyonel Tasarruf', 'Müşteri Memnuniyeti'],
        values=[artan_donusum_geliri, operasyonel_tasarruf, musteri_memnuniyeti_degeri],
        marker_colors=['#BAFFC9', '#BAE1FF', '#E0BBE4'],
        hole=0.3
    ),
    row=1, col=2
)

# 3. Maliyet vs Fayda (Bar Chart)
fig.add_trace(
    go.Bar(
        x=['Maliyet', 'Fayda', 'Net Kazanç'],
        y=[toplam_maliyet, toplam_fayda, net_fayda],
        marker_color=['#FFB3BA', '#BAFFC9', '#BAE1FF'],
        text=[f'€{toplam_maliyet:,}', f'€{toplam_fayda:,}', f'€{net_fayda:,}'],
        textposition='outside'
    ),
    row=2, col=1
)

# 4. ROI Göstergesi (Indicator)
fig.add_trace(
    go.Indicator(
        mode="gauge+number+delta",
        value=roi,
        title={'text': "ROI (%)"},
        delta={'reference': 100, 'increasing': {'color': "green"}},
        gauge={
            'axis': {'range': [0, 400]},
            'bar': {'color': "#34A853"},
            'steps': [
                {'range': [0, 100], 'color': "#FFE6E6"},
                {'range': [100, 200], 'color': "#FFF9E6"},
                {'range': [200, 400], 'color': "#E6F4EA"}
            ],
            'threshold': {
                'line': {'color': "red", 'width': 4},
                'thickness': 0.75,
                'value': 200
            }
        }
    ),
    row=2, col=2
)

fig.update_layout(
    title_text="Proje Maliyet-Fayda Analizi",
    height=800,
    showlegend=True
)

fig.show()

print("\n✅ SONUÇ:")
print("="*60)
if roi > 200:
    print("🎯 Proje son derece karlı! Yatırım öneriliyor.")
elif roi > 100:
    print("✅ Proje karlı. Yatırım yapılabilir.")
else:
    print("⚠️ Proje riskli. Ek analize ihtiyaç var.")
print("="*60)

💰 MALİYET-FAYDA ANALİZİ SONUÇLARI
📉 Toplam Proje Maliyeti    : €35,000
📈 Yıllık Toplam Fayda      : €145,000
💵 Net Fayda (İlk Yıl)      : €110,000
📊 ROI (Yatırım Getirisi)   : %314.3
⏱️  Geri Ödeme Süresi        : 2.9 ay (~0.24 yıl)



✅ SONUÇ:
🎯 Proje son derece karlı! Yatırım öneriliyor.


### ⚠️ **Risk Analizi ve Azaltma Stratejileri**

**YBS yaklaşımı:** Her projede riskleri önceden belirlemek ve plan yapmak önemlidir.

| Risk | Olasılık | Etki | Azaltma Stratejisi |
|------|----------|------|-------------------|
| **Veri Kalitesi Sorunları** | Orta | Yüksek | Detaylı veri temizleme, validasyon |
| **Model Performansı Düşük** | Düşük | Yüksek | Cross-validation, hiperparametre optimizasyonu |
| **Sistem Entegrasyonu Gecikmesi** | Orta | Orta | Erken IT görüşmeleri, API tasarımı |
| **Kullanıcı Direnci** | Orta | Orta | Eğitim programları, pilot uygulama |
| **Yasal/Etik Sorunlar (KVKK)** | Düşük | Yüksek | Hukuk departmanı onayı, anonimizasyon |
| **Model Güncelliği** | Yüksek | Orta | Otomatik model retraining sistemi |




```mermaid
graph LR
    A[Risk Yönetimi] --> B[Tespit Et]
    B --> C[Değerlendir]
    C --> D[Önlem Al]
    D --> E[İzle]
    E --> B
    
    style A fill:#FFE5CC,stroke:#D68910,stroke-width:3px,color:#000
    style B fill:#FFE6E6,stroke:#C0392B,stroke-width:2px,color:#000
    style C fill:#FFF9E6,stroke:#F39C12,stroke-width:2px,color:#000
    style D fill:#E6F4EA,stroke:#34A853,stroke-width:2px,color:#000
    style E fill:#E8F4F8,stroke:#2C5F7C,stroke-width:2px,color:#000
```

### 🤖 **İş Probleminden ML Problemine Dönüşüm**

**Kritik adım:** İş problemini makine öğrenmesi problemi olarak formüle etmek.

#### 📋 **Problem Formülasyonu**

**İş Problemi:**
> "Telefon pazarlama kampanyasında hangi müşterilere öncelik vermeliyiz?"

**↓ DÖNÜŞÜM ↓**

**ML Problemi:**
> **İkili Sınıflandırma (Binary Classification):**  
> Müşteri özelliklerine (demografik, finansal, kampanya geçmişi) bakarak, müşteriinin vadeli mevduat açıp açmayacağını tahmin etmek.

```mermaid
graph TB
    A[İş Problemi<br/>Kampanya Verimliliği] --> B{ML Problem Tipi}
    
    B --> C[Sınıflandırma ✅<br/>Classification]
    B --> D[Regresyon ❌<br/>Regression]
    B --> E[Kümeleme ❌<br/>Clustering]
    
    C --> F[İkili Sınıflandırma ✅<br/>Binary Classification]
    C --> G[Çoklu Sınıflandırma ❌<br/>Multi-class]
    
    F --> H[Lojistik Regresyon<br/>Model Seçimi]
    
    H --> I[Input: Müşteri Özellikleri<br/>Output: Yes/No Tahmini<br/>Metrik: Accuracy, Recall, F1]
    
    style A fill:#FFE5CC,stroke:#D68910,stroke-width:3px,color:#000
    style B fill:#FFF9E6,stroke:#F39C12,stroke-width:2px,color:#000
    style C fill:#BAFFC9,stroke:#28A745,stroke-width:3px,color:#000
    style D fill:#FFE6E6,stroke:#C0392B,stroke-width:1px,color:#000
    style E fill:#FFE6E6,stroke:#C0392B,stroke-width:1px,color:#000
    style F fill:#E6F4EA,stroke:#34A853,stroke-width:3px,color:#000
    style G fill:#FFE6E6,stroke:#C0392B,stroke-width:1px,color:#000
    style H fill:#F4E6FF,stroke:#8E44AD,stroke-width:3px,color:#000
    style I fill:#E8F4F8,stroke:#2C5F7C,stroke-width:2px,color:#000
```

#### 🎯 **ML Problem Tanımı**

| Bileşen | Tanım |
|---------|-------|
| **Problem Tipi** | Supervised Learning - Binary Classification |
| **Input (X)** | 20 özellik (age, job, balance, campaign, vb.) |
| **Output (y)** | 0 (Almayacak) veya 1 (Alacak) |
| **Algoritma** | Lojistik Regresyon (başlangıç için) |
| **Başarı Metrikleri** | Accuracy, Precision, Recall, F1-Score, AUC-ROC |
| **Kısıtlamalar** | - Model yorumlanabilir olmalı<br/>- Gerçek zamanlı tahmin yapabilmeli<br/>- %85+ accuracy hedefi |

### 🎓 YBS Perspektifinden Önemli Çıkarımlar

> **💡 Ders 1:** Makine öğrenmesi projeleri teknolojiyle değil, **iş problemiyle** başlar. Teknik mükemmellik yeterli değildir; iş değeri yaratmalısınız.

> **💡 Ders 2:** Stakeholder'ları anlamak ve yönetmek, teknik yetenekler kadar önemlidir. Her paydaşın farklı beklentileri vardır.

> **💡 Ders 3:** ROI hesaplaması yapmadan proje başlatmayın. Üst yönetime somut sayılarla değer göstermelisiniz.

> **💡 Ders 4:** Riskleri önceden belirleyin ve azaltma planları yapın. "Beklenmedik" sorunlar, aslında önceden görülebilir.

> **💡 Ders 5:** İş problemini ML problemine dönüştürürken, doğru problem tipini seçmek başarının yarısıdır.

## 🔍 **Adım 4: Data Understanding - Keşifsel Veri Analizi (EDA)**

**CRISP-DM'nin ikinci aşaması:** Veriyi derinlemesine anlamak, kalite sorunlarını tespit etmek ve içgörüler elde etmek.

> **💡 YBS Notu:** Bu aşamada veriyi iş problemini çözmeye yönelik olarak analiz ediyoruz. Sadece istatistik değil, iş anlayışı da gereklidir.

### 📊 **Veri Yapısı ve Değişken Tipleri**

EDA sürecinin ilk adımı: Veri tiplerini anlamak ve kategorilere ayırmak.

In [23]:
# Değişkenleri kategorilere ayırma
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_cols = df.select_dtypes(include=['object']).columns.tolist()

# Hedef değişkeni kategorik listeden çıkar
if 'y' in categorical_cols:
    categorical_cols.remove('y')

print("📊 DEĞİŞKEN TİPLERİ ANALİZİ")
print("="*70)
print(f"📏 Toplam Değişken Sayısı       : {df.shape[1]}")
print(f"🔢 Numerik Değişken Sayısı      : {len(numerical_cols)}")
print(f"📝 Kategorik Değişken Sayısı    : {len(categorical_cols)}")
print(f"🎯 Hedef Değişken                : y (kategorik)")
print("="*70)

print("\n🔢 NUMERIK DEĞİŞKENLER:")
print("-" * 70)
for i, col in enumerate(numerical_cols, 1):
    print(f"{i:2d}. {col:20s} | Tip: {str(df[col].dtype):10s} | Unique: {df[col].nunique()}")

print("\n📝 KATEGORİK DEĞİŞKENLER:")
print("-" * 70)
for i, col in enumerate(categorical_cols, 1):
    print(f"{i:2d}. {col:20s} | Tip: {str(df[col].dtype):10s} | Unique: {df[col].nunique()}")

print("\n🎯 HEDEF DEĞİŞKEN:")
print("-" * 70)
print(f"    y                    | Tip: {str(df['y'].dtype):10s} | Unique: {df['y'].nunique()}")
print("="*70)

📊 DEĞİŞKEN TİPLERİ ANALİZİ
📏 Toplam Değişken Sayısı       : 21
🔢 Numerik Değişken Sayısı      : 11
📝 Kategorik Değişken Sayısı    : 10
🎯 Hedef Değişken                : y (kategorik)

🔢 NUMERIK DEĞİŞKENLER:
----------------------------------------------------------------------
 1. age                  | Tip: int64      | Unique: 78
 2. duration             | Tip: int64      | Unique: 1544
 3. campaign             | Tip: int64      | Unique: 42
 4. pdays                | Tip: int64      | Unique: 27
 5. previous             | Tip: int64      | Unique: 8
 6. emp_var_rate         | Tip: float64    | Unique: 10
 7. cons_price_idx       | Tip: float64    | Unique: 26
 8. cons_conf_idx        | Tip: float64    | Unique: 26
 9. euribor3m            | Tip: float64    | Unique: 316
10. nr_employed          | Tip: float64    | Unique: 11
11. y                    | Tip: int64      | Unique: 2

📝 KATEGORİK DEĞİŞKENLER:
----------------------------------------------------------------------
 1. job 

### 🔍 **Eksik Veri Analizi**

Eksik veriler (missing values/NaN) modelin performansını olumsuz etkileyebilir. Bu yüzden veri setindeki eksik değerleri tespit etmek ve uygun şekilde ele almak kritik öneme sahiptir.

In [25]:
# Eksik veri analizi
missing_data = pd.DataFrame({
    'Değişken': df.columns,
    'Eksik Sayı': df.isnull().sum(),
    'Eksik Yüzde (%)': (df.isnull().sum() / len(df) * 100).round(2)
}).sort_values('Eksik Sayı', ascending=False)

print("🔍 EKSİK VERİ ANALİZİ")
print("="*70)
print(f"📊 Toplam Kayıt Sayısı          : {len(df):,}")
print(f"🔢 Toplam Değişken Sayısı       : {df.shape[1]}")
print(f"❌ Eksik Veri İçeren Değişken   : {(missing_data['Eksik Sayı'] > 0).sum()}")
print("="*70)

if missing_data['Eksik Sayı'].sum() == 0:
    print("\n✅ TEBRİKLER! Veri setinde hiç eksik veri yok!")
    print("   Bu, veri kalitesinin yüksek olduğunu gösterir.")
else:
    print("\n⚠️  EKSİK VERİ DETAYLARI:")
    print("-" * 70)
    print(missing_data[missing_data['Eksik Sayı'] > 0].to_string(index=False))

print("\n" + "="*70)

🔍 EKSİK VERİ ANALİZİ
📊 Toplam Kayıt Sayısı          : 41,188
🔢 Toplam Değişken Sayısı       : 21
❌ Eksik Veri İçeren Değişken   : 0

✅ TEBRİKLER! Veri setinde hiç eksik veri yok!
   Bu, veri kalitesinin yüksek olduğunu gösterir.



### 📈 **Numerik Değişkenlerin İstatistiksel Özeti**

Numerik değişkenlerin merkezi eğilim (ortalama, medyan) ve dağılım ölçülerini (standart sapma, min, max) inceleyeceğiz.

In [27]:
# Hedef değişkeni (y) numeric listeden çıkar (zaten kategorik olarak işliyoruz)
numerical_features = [col for col in numerical_cols if col != 'y']

# İstatistiksel özet
print("📈 NUMERİK DEĞİŞKENLER - İSTATİSTİKSEL ÖZET")
print("="*100)
stats_summary = df[numerical_features].describe().T
stats_summary['skewness'] = df[numerical_features].skew()
stats_summary['kurtosis'] = df[numerical_features].kurtosis()

# Türkçe sütun başlıkları ile göster
stats_summary_display = stats_summary.copy()
stats_summary_display.columns = ['Gözlem', 'Ortalama', 'Std. Sapma', 'Min', '25%', '50% (Medyan)', '75%', 'Max', 'Çarpıklık', 'Basıklık']

print(stats_summary_display.round(2))
print("="*100)

print("\n💡 İSTATİSTİKSEL YORUMLAR:")
print("-" * 100)
print("   • Çarpıklık (Skewness): -2 ile +2 arası normal kabul edilir")
print("   • Basıklık (Kurtosis): Normal dağılım için 3 civarında olmalıdır")
print("   • Standart Sapma: Veri dağılımının ne kadar geniş olduğunu gösterir")
print("="*100)

📈 NUMERİK DEĞİŞKENLER - İSTATİSTİKSEL ÖZET
                 Gözlem  Ortalama  Std. Sapma      Min      25%  50% (Medyan)  \
age             41188.0     40.02       10.42    17.00    32.00         38.00   
duration        41188.0    258.29      259.28     0.00   102.00        180.00   
campaign        41188.0      2.57        2.77     1.00     1.00          2.00   
pdays           41188.0    962.48      186.91     0.00   999.00        999.00   
previous        41188.0      0.17        0.49     0.00     0.00          0.00   
emp_var_rate    41188.0      0.08        1.57    -3.40    -1.80          1.10   
cons_price_idx  41188.0     93.58        0.58    92.20    93.08         93.75   
cons_conf_idx   41188.0    -40.50        4.63   -50.80   -42.70        -41.80   
euribor3m       41188.0      3.62        1.73     0.63     1.34          4.86   
nr_employed     41188.0   5167.04       72.25  4963.60  5099.10       5191.00   

                    75%      Max  Çarpıklık  Basıklık  
age      

### 📝 **Kategorik Değişkenlerin Analizi**

Kategorik değişkenlerin benzersiz değer sayıları ve en sık görülen kategorilerini inceleyelim.

In [29]:
# Kategorik değişkenlerin analizi
print("📝 KATEGORİK DEĞİŞKENLER - DETAYLI ANALİZ")
print("="*100)

for col in categorical_cols:
    print(f"\n🔹 {col.upper()}")
    print("-" * 100)
    print(f"   Benzersiz Kategori Sayısı: {df[col].nunique()}")
    print(f"   En Sık Görülen Değer     : {df[col].mode()[0]} ({df[col].value_counts().iloc[0]:,} adet)")
    print(f"\n   Değer Dağılımı:")
    
    value_counts = df[col].value_counts()
    value_percentages = (df[col].value_counts(normalize=True) * 100).round(2)
    
    for idx, (val, count) in enumerate(value_counts.head(5).items(), 1):
        percent = value_percentages[val]
        bar = "█" * int(percent / 2)  # Her %2 için bir blok
        print(f"      {idx}. {str(val):20s}: {count:7,} ({percent:6.2f}%)  {bar}")
    
    if len(value_counts) > 5:
        print(f"      ... ve {len(value_counts) - 5} kategori daha")

print("\n" + "="*100)

📝 KATEGORİK DEĞİŞKENLER - DETAYLI ANALİZ

🔹 JOB
----------------------------------------------------------------------------------------------------
   Benzersiz Kategori Sayısı: 12
   En Sık Görülen Değer     : admin. (10,422 adet)

   Değer Dağılımı:
      1. admin.              :  10,422 ( 25.30%)  ████████████
      2. blue-collar         :   9,254 ( 22.47%)  ███████████
      3. technician          :   6,743 ( 16.37%)  ████████
      4. services            :   3,969 (  9.64%)  ████
      5. management          :   2,924 (  7.10%)  ███
      ... ve 7 kategori daha

🔹 MARITAL
----------------------------------------------------------------------------------------------------
   Benzersiz Kategori Sayısı: 4
   En Sık Görülen Değer     : married (24,928 adet)

   Değer Dağılımı:
      1. married             :  24,928 ( 60.52%)  ██████████████████████████████
      2. single              :  11,568 ( 28.09%)  ██████████████
      3. divorced            :   4,612 ( 11.20%)  █████
      4

### 📊 **Numerik Değişkenlerin Görselleştirilmesi**

Numerik değişkenlerin dağılımlarını (histogram) ve kutu grafiklerini (box plot) inceleyerek aykırı değerleri ve veri dağılım şekillerini görselleştirelim.

In [31]:
# Numerik değişkenlerin histogramları
from plotly.subplots import make_subplots
import plotly.graph_objects as go

# İnteraktif histogramlar oluştur
fig = make_subplots(
    rows=5, cols=2,
    subplot_titles=[col for col in numerical_features],
    vertical_spacing=0.08,
    horizontal_spacing=0.12
)

# Pastel renkler paleti
colors = ['#FFB6C1', '#B0E0E6', '#FFE4B5', '#98FB98', '#DDA0DD', 
          '#F0E68C', '#FFDAB9', '#E0BBFF', '#FFCCFF', '#C1FFC1']

for idx, col in enumerate(numerical_features):
    row = idx // 2 + 1
    col_pos = idx % 2 + 1
    
    fig.add_trace(
        go.Histogram(
            x=df[col],
            name=col,
            marker=dict(color=colors[idx], line=dict(color='white', width=1)),
            showlegend=False
        ),
        row=row, col=col_pos
    )

# Layout ayarları
fig.update_layout(
    height=1200,
    title_text="<b>Numerik Değişkenlerin Dağılımı (Histogramlar)</b>",
    title_font=dict(size=18, color='#2F4F4F'),
    showlegend=False,
    plot_bgcolor='white',
    paper_bgcolor='#F5F5F5'
)

# Eksen ayarları
fig.update_xaxes(showgrid=True, gridcolor='lightgray', zeroline=False)
fig.update_yaxes(showgrid=True, gridcolor='lightgray', zeroline=False)

fig.show()

print("✅ Histogramlar başarıyla oluşturuldu!")

✅ Histogramlar başarıyla oluşturuldu!


In [33]:
# Numerik değişkenlerin kutu grafikleri (Box Plots)
fig = make_subplots(
    rows=5, cols=2,
    subplot_titles=[col for col in numerical_features],
    vertical_spacing=0.08,
    horizontal_spacing=0.12
)

for idx, col in enumerate(numerical_features):
    row = idx // 2 + 1
    col_pos = idx % 2 + 1
    
    fig.add_trace(
        go.Box(
            y=df[col],
            name=col,
            marker=dict(color=colors[idx]),
            boxmean='sd',  # Ortalama ve standart sapma göster
            showlegend=False
        ),
        row=row, col=col_pos
    )

# Layout ayarları
fig.update_layout(
    height=1200,
    title_text="<b>Numerik Değişkenlerin Kutu Grafikleri (Aykırı Değer Analizi)</b>",
    title_font=dict(size=18, color='#2F4F4F'),
    showlegend=False,
    plot_bgcolor='white',
    paper_bgcolor='#F5F5F5'
)

# Eksen ayarları
fig.update_xaxes(showgrid=True, gridcolor='lightgray')
fig.update_yaxes(showgrid=True, gridcolor='lightgray', zeroline=False)

fig.show()

print("✅ Kutu grafikleri başarıyla oluşturuldu!")
print("\n💡 Kutu Grafiği Yorumlama:")
print("   • Kutunun içindeki çizgi: Medyan (50. yüzdelik)")
print("   • Kutunun alt sınırı: 25. yüzdelik (Q1)")
print("   • Kutunun üst sınırı: 75. yüzdelik (Q3)")
print("   • Bıyıklar: Q1-1.5*IQR ve Q3+1.5*IQR")
print("   • Noktalar: Aykırı değerler (outliers)")

✅ Kutu grafikleri başarıyla oluşturuldu!

💡 Kutu Grafiği Yorumlama:
   • Kutunun içindeki çizgi: Medyan (50. yüzdelik)
   • Kutunun alt sınırı: 25. yüzdelik (Q1)
   • Kutunun üst sınırı: 75. yüzdelik (Q3)
   • Bıyıklar: Q1-1.5*IQR ve Q3+1.5*IQR
   • Noktalar: Aykırı değerler (outliers)


### 🔗 **Korelasyon Analizi**

Korelasyon analizi, değişkenler arasındaki ilişkinin gücünü ve yönünü ölçer. Pearson korelasyon katsayısı **-1 ile +1** arasında değer alır:

- **+1**: Mükemmel pozitif ilişki
- **0**: İlişki yok  
- **-1**: Mükemmel negatif ilişki

Makine öğrenmesinde **yüksek korelasyonlu değişkenler** (multicollinearity) model performansını olumsuz etkileyebilir.

In [37]:
# Korelasyon matrisi hesaplama
correlation_matrix = df[numerical_features + ['y']].corr()

# İnteraktif korelasyon ısı haritası (heatmap)
fig = go.Figure(data=go.Heatmap(
    z=correlation_matrix.values,
    x=correlation_matrix.columns,
    y=correlation_matrix.columns,
    colorscale='RdYlGn',  # Kırmızı-Sarı-Yeşil
    zmid=0,  # Orta nokta 0'da
    text=correlation_matrix.round(2).values,
    texttemplate='%{text}',
    textfont={"size": 10},
    colorbar=dict(
        title=dict(text="Korelasyon", side="right"),
        tickmode="linear",
        tick0=-1,
        dtick=0.5
    )
))

fig.update_layout(
    title="<b>Numerik Değişkenler ve Hedef Değişken Korelasyon Matrisi</b>",
    title_font=dict(size=18, color='#2F4F4F'),
    width=900,
    height=800,
    xaxis=dict(tickangle=-45),
    plot_bgcolor='white',
    paper_bgcolor='#F5F5F5'
)

fig.show()

print("✅ Korelasyon matrisi başarıyla oluşturuldu!")
print("\n📊 HEDEF DEĞİŞKEN (y) İLE EN YÜKSEK KORELASYONLAR:")
print("="*70)

# Hedef değişken ile korelasyonları sırala
target_corr = correlation_matrix['y'].drop('y').abs().sort_values(ascending=False)

for idx, (feature, corr_value) in enumerate(target_corr.items(), 1):
    original_corr = correlation_matrix['y'][feature]
    direction = "Pozitif ↑" if original_corr > 0 else "Negatif ↓"
    bar = "█" * int(abs(original_corr) * 50)  # Görsel bar
    print(f"{idx:2d}. {feature:20s}: {original_corr:7.4f} ({direction:12s}) {bar}")

✅ Korelasyon matrisi başarıyla oluşturuldu!

📊 HEDEF DEĞİŞKEN (y) İLE EN YÜKSEK KORELASYONLAR:
 1. duration            :  0.4053 (Pozitif ↑   ) ████████████████████
 2. nr_employed         : -0.3547 (Negatif ↓   ) █████████████████
 3. pdays               : -0.3249 (Negatif ↓   ) ████████████████
 4. euribor3m           : -0.3078 (Negatif ↓   ) ███████████████
 5. emp_var_rate        : -0.2983 (Negatif ↓   ) ██████████████
 6. previous            :  0.2302 (Pozitif ↑   ) ███████████
 7. cons_price_idx      : -0.1362 (Negatif ↓   ) ██████
 8. campaign            : -0.0664 (Negatif ↓   ) ███
 9. cons_conf_idx       :  0.0549 (Pozitif ↑   ) ██
10. age                 :  0.0304 (Pozitif ↑   ) █


### 🎯 **Hedef Değişken ile İlişki Analizi (Bivariate Analysis)**

Hedef değişken (y) ile diğer kategorik değişkenler arasındaki ilişkiyi görselleştirerek hangi kategorilerin başarı oranını etkilediğini inceleyelim.

In [39]:
# Hedef değişken ile kategorik değişkenler arasındaki ilişki
# En ilginç 6 kategorik değişkeni seçelim
key_categorical = ['job', 'marital', 'education', 'contact', 'month', 'poutcome']

fig = make_subplots(
    rows=3, cols=2,
    subplot_titles=key_categorical,
    vertical_spacing=0.12,
    horizontal_spacing=0.12
)

for idx, col in enumerate(key_categorical):
    row = idx // 2 + 1
    col_pos = idx % 2 + 1
    
    # Crosstab ile ilişki analizi
    ct = pd.crosstab(df[col], df['y'], normalize='index') * 100
    
    # Her kategori için yes (1) oranını al
    categories = ct.index.tolist()
    yes_rates = ct[1].tolist()
    
    fig.add_trace(
        go.Bar(
            x=categories,
            y=yes_rates,
            name=col,
            marker=dict(color=colors[idx % len(colors)]),
            showlegend=False,
            text=[f'{val:.1f}%' for val in yes_rates],
            textposition='outside'
        ),
        row=row, col=col_pos
    )
    
    # Y ekseni başlığı
    fig.update_yaxes(title_text="Başarı Oranı (%)", row=row, col=col_pos)

# Layout ayarları
fig.update_layout(
    height=900,
    title_text="<b>Kategorik Değişkenlere Göre Hedef Değişken Başarı Oranları</b>",
    title_font=dict(size=18, color='#2F4F4F'),
    showlegend=False,
    plot_bgcolor='white',
    paper_bgcolor='#F5F5F5'
)

# X ekseni label rotation
fig.update_xaxes(tickangle=-45)
fig.update_yaxes(showgrid=True, gridcolor='lightgray')

fig.show()

print("✅ Bivariate analiz grafikleri başarıyla oluşturuldu!")
print("\n💡 YORUMLAMA:")
print("   • Yüksek başarı oranı: İlgili kategori pozitif sonuç için önemli")
print("   • Düşük başarı oranı: İlgili kategori negatif etkili olabilir")
print("   • Bu grafikler feature engineering için ipuçları verir")

✅ Bivariate analiz grafikleri başarıyla oluşturuldu!

💡 YORUMLAMA:
   • Yüksek başarı oranı: İlgili kategori pozitif sonuç için önemli
   • Düşük başarı oranı: İlgili kategori negatif etkili olabilir
   • Bu grafikler feature engineering için ipuçları verir


### 🔑 **EDA'dan Önemli Bulgular**

Keşifsel Veri Analizi sonucunda elde ettiğimiz önemli bulgular:

#### **Veri Kalitesi**
- ✅ **Eksik veri yok**: 41,188 kayıt tamamen dolu
- ✅ **Yeterli veri büyüklüğü**: Makine öğrenmesi için yeterli örneklem

#### **Hedef Değişken (y)**
- ⚠️ **Dengesiz dağılım**: %88.3 hayır, %11.7 evet
- Bu durum **class imbalance** problemi oluşturabilir
- Çözüm: SMOTE, class_weight veya threshold ayarı kullanılabilir

#### **En Etkili Numerik Değişkenler**
1. **duration** (görüşme süresi): +0.405 korelasyon ⭐
2. **nr_employed** (çalışan sayısı): -0.355 korelasyon
3. **pdays** (son temasa geçen gün): -0.325 korelasyon
4. **euribor3m** (faiz oranı): -0.308 korelasyon

#### **Kategorik Değişkenlerden Önemli Bulgular**
- **poutcome** (önceki kampanya sonucu): "success" yüksek başarı oranı
- **contact** (iletişim türü): "cellular" daha etkili
- **month** (ay): Mart, Eylül, Ekim, Aralık ayları daha başarılı
- **job** (meslek): Öğrenci ve emekliler daha yüksek dönüşüm oranı

#### **Veri Hazırlama İçin Öneriler**
1. **Duration** değişkeni: Gerçek hayat tahminlerinde bilinmez, dikkatli kullanılmalı
2. **Scaling gerekli**: Numerik değişkenlerin ölçekleri çok farklı
3. **Encoding gerekli**: Kategorik değişkenler sayısal değerlere dönüştürülmeli
4. **Multicollinearity kontrol**: euribor3m, emp_var_rate, nr_employed yüksek korelasyonlu
5. **Outlier yönetimi**: Bazı değişkenlerde (duration, campaign) aykırı değerler mevcut

## 🔧 **Adım 5: Data Preparation - Veri Hazırlama**

Veri hazırlama, ham veriyi makine öğrenmesi modellerine uygun hale getirme sürecidir. Bu aşamada:

- **Kategorik değişkenleri** sayısal değerlere dönüştüreceğiz (Encoding)
- **Numerik değişkenleri** ölçeklendireceğiz (Scaling)
- Veriyi **eğitim ve test** setlerine ayıracağız
- **Class imbalance** sorununu ele alacağız

Bu adımlar, modelin doğru ve tutarlı öğrenmesi için kritik öneme sahiptir.

### 🔄 **Hedef Değişkeni Ayırma**

İlk adım olarak hedef değişkeni (y) bağımsız değişkenlerden (X) ayırmalıyız.

In [41]:
# Hedef değişken ve bağımsız değişkenleri ayır
X = df.drop('y', axis=1)
y = df['y']

print("🎯 HEDEF DEĞİŞKEN VE BAĞIMSIZ DEĞİŞKENLER AYRILDI")
print("="*70)
print(f"📊 Bağımsız Değişkenler (X) Boyutu: {X.shape}")
print(f"🎯 Hedef Değişken (y) Boyutu      : {y.shape}")
print(f"📝 Toplam Özellik Sayısı          : {X.shape[1]}")
print("="*70)

print("\n📋 X (Bağımsız Değişkenler) - İlk 5 satır:")
print(X.head())

🎯 HEDEF DEĞİŞKEN VE BAĞIMSIZ DEĞİŞKENLER AYRILDI
📊 Bağımsız Değişkenler (X) Boyutu: (41188, 20)
🎯 Hedef Değişken (y) Boyutu      : (41188,)
📝 Toplam Özellik Sayısı          : 20

📋 X (Bağımsız Değişkenler) - İlk 5 satır:
   age          job  marital          education  default housing loan  \
0   44  blue-collar  married           basic.4y  unknown     yes   no   
1   53   technician  married            unknown       no      no   no   
2   28   management   single  university.degree       no     yes   no   
3   39     services  married        high.school       no      no   no   
4   55      retired  married           basic.4y       no     yes   no   

    contact month day_of_week  duration  campaign  pdays  previous  \
0  cellular   aug         thu       210         1    999         0   
1  cellular   nov         fri       138         1    999         0   
2  cellular   jun         thu       339         3      6         2   
3  cellular   apr         fri       185         2    999    

### 🔤 **Kategorik Değişkenlerin Encoding İşlemi**

Makine öğrenmesi algoritmaları sadece sayısal verilerle çalışabilir. Bu nedenle kategorik değişkenleri sayısal değerlere dönüştürmeliyiz.

**Encoding Yöntemleri:**

1. **Label Encoding**: Kategorileri 0, 1, 2... şeklinde sayılara çevirir (sıralı kategoriler için)
2. **One-Hot Encoding**: Her kategori için yeni bir binary (0/1) sütun oluşturur (sırasız kategoriler için)

Bu projede **One-Hot Encoding** kullanacağız çünkü kategorilerimiz arasında doğal bir sıralama yok.

In [43]:
# Kategorik değişkenleri tespit et (X'ten)
categorical_features = X.select_dtypes(include=['object']).columns.tolist()
numerical_features_X = X.select_dtypes(include=['int64', 'float64']).columns.tolist()

print("🔤 KATEGORİK DEĞİŞKENLER - ONE-HOT ENCODING")
print("="*70)
print(f"📝 Kategorik Değişken Sayısı: {len(categorical_features)}")
print("   Kategorik Değişkenler:", categorical_features)
print(f"\n🔢 Numerik Değişken Sayısı: {len(numerical_features_X)}")
print("="*70)

# One-Hot Encoding uygula
X_encoded = pd.get_dummies(X, columns=categorical_features, drop_first=True, dtype=int)

print(f"\n✅ ONE-HOT ENCODING TAMAMLANDI!")
print("="*70)
print(f"📊 Encoding Öncesi Özellik Sayısı : {X.shape[1]}")
print(f"📊 Encoding Sonrası Özellik Sayısı: {X_encoded.shape[1]}")
print(f"➕ Eklenen Özellik Sayısı         : {X_encoded.shape[1] - X.shape[1]}")
print("="*70)

print("\n📋 X_encoded (Encoded Veriler) - İlk 5 satır, ilk 15 sütun:")
print(X_encoded.iloc[:5, :15])

🔤 KATEGORİK DEĞİŞKENLER - ONE-HOT ENCODING
📝 Kategorik Değişken Sayısı: 10
   Kategorik Değişkenler: ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome']

🔢 Numerik Değişken Sayısı: 10

✅ ONE-HOT ENCODING TAMAMLANDI!
📊 Encoding Öncesi Özellik Sayısı : 20
📊 Encoding Sonrası Özellik Sayısı: 53
➕ Eklenen Özellik Sayısı         : 33

📋 X_encoded (Encoded Veriler) - İlk 5 satır, ilk 15 sütun:
   age  duration  campaign  pdays  previous  emp_var_rate  cons_price_idx  \
0   44       210         1    999         0           1.4          93.444   
1   53       138         1    999         0          -0.1          93.200   
2   28       339         3      6         2          -1.7          94.055   
3   39       185         2    999         0          -1.8          93.075   
4   55       137         1      3         1          -2.9          92.201   

   cons_conf_idx  euribor3m  nr_employed  job_blue-collar  job_entrepreneur  \
0          

### ✂️ **Train-Test Split (Veri Bölme)**

Makine öğrenmesinde veriyi **eğitim (train)** ve **test** setlerine ayırmak çok önemlidir:

- **Eğitim Seti (Training Set)**: Model bu veriden öğrenir
- **Test Seti (Test Set)**: Modelin performansını değerlendirmek için kullanılır (modelin görmediği veri)

**Stratified Split**: Hedef değişkenin dağılımını koruyarak bölme yapar. Class imbalance durumlarında kritik öneme sahiptir.

**Oran**: Genellikle 80% eğitim, 20% test kullanılır.

In [45]:
from sklearn.model_selection import train_test_split

# Stratified split ile veriyi ayır (hedef değişkenin dağılımını koru)
X_train, X_test, y_train, y_test = train_test_split(
    X_encoded, 
    y, 
    test_size=0.2,      # %20 test
    random_state=42,    # Tekrar edilebilirlik için
    stratify=y          # Hedef değişken dağılımını koru
)

print("✂️ TRAIN-TEST SPLIT TAMAMLANDI")
print("="*70)
print(f"📊 Toplam Veri Sayısı             : {len(X_encoded):,}")
print(f"📚 Eğitim Seti Boyutu (80%)       : {X_train.shape}")
print(f"🧪 Test Seti Boyutu (20%)         : {X_test.shape}")
print("="*70)

# Hedef değişkenin dağılımını kontrol et
train_dist = y_train.value_counts(normalize=True) * 100
test_dist = y_test.value_counts(normalize=True) * 100

print("\n🎯 HEDEF DEĞİŞKEN DAĞILIMI:")
print("-"*70)
print(f"   Eğitim Seti  - 0 (Hayır): {train_dist[0]:.2f}% | 1 (Evet): {train_dist[1]:.2f}%")
print(f"   Test Seti    - 0 (Hayır): {test_dist[0]:.2f}% | 1 (Evet): {test_dist[1]:.2f}%")
print("-"*70)
print("   ✅ Stratified split sayesinde dağılım korundu!")
print("="*70)

✂️ TRAIN-TEST SPLIT TAMAMLANDI
📊 Toplam Veri Sayısı             : 41,188
📚 Eğitim Seti Boyutu (80%)       : (32950, 53)
🧪 Test Seti Boyutu (20%)         : (8238, 53)

🎯 HEDEF DEĞİŞKEN DAĞILIMI:
----------------------------------------------------------------------
   Eğitim Seti  - 0 (Hayır): 88.73% | 1 (Evet): 11.27%
   Test Seti    - 0 (Hayır): 88.74% | 1 (Evet): 11.26%
----------------------------------------------------------------------
   ✅ Stratified split sayesinde dağılım korundu!


### ⚖️ **Feature Scaling (Özellik Ölçeklendirme)**

Farklı ölçeklerdeki değişkenler (örn: yaş 18-98, nr_employed 4963-5228) modelin öğrenmesini olumsuz etkileyebilir. Ölçeklendirme, tüm özellikleri benzer bir aralığa getirir.

**Scaling Yöntemleri:**

1. **StandardScaler**: Verileri ortalama=0, standart sapma=1 olacak şekilde dönüştürür
   - Formül: z = (x - μ) / σ
   - Outlier'lara duyarlı
   
2. **MinMaxScaler**: Verileri [0, 1] aralığına dönüştürür
   - Formül: x' = (x - min) / (max - min)
   - Outlier'lara çok duyarlı

**Lojistik Regresyon** için genellikle **StandardScaler** tercih edilir.

**ÖNEMLİ**: Scaling işlemi SADECE eğitim verisi üzerinden öğrenilmeli, test verisine uygulanmalı (data leakage'i önlemek için).

In [47]:
from sklearn.preprocessing import StandardScaler

# StandardScaler oluştur
scaler = StandardScaler()

# Sadece eğitim verisinden öğren ve dönüştür
X_train_scaled = scaler.fit_transform(X_train)

# Test verisini eğitim setinden öğrenilen parametrelerle dönüştür
X_test_scaled = scaler.transform(X_test)

print("⚖️ FEATURE SCALING TAMAMLANDI")
print("="*70)
print(f"📚 X_train_scaled Boyutu: {X_train_scaled.shape}")
print(f"🧪 X_test_scaled Boyutu : {X_test_scaled.shape}")
print("="*70)

# Ölçeklendirme öncesi ve sonrası karşılaştırma
print("\n📊 ÖLÇEKLENDİRME ÖNCESİ VE SONRASI KARŞILAŞTIRMA:")
print("-"*70)
print("   İlk 5 özellik için istatistikler:\n")

comparison_df = pd.DataFrame({
    'Özellik': X_train.columns[:5],
    'Önce Ort.': X_train.iloc[:, :5].mean().round(2).values,
    'Önce Std.': X_train.iloc[:, :5].std().round(2).values,
    'Sonra Ort.': X_train_scaled[:, :5].mean(axis=0).round(2),
    'Sonra Std.': X_train_scaled[:, :5].std(axis=0).round(2)
})

print(comparison_df.to_string(index=False))
print("-"*70)
print("   ✅ Tüm özellikler ortalama=0, std=1 olarak ölçeklendirildi!")
print("="*70)

⚖️ FEATURE SCALING TAMAMLANDI
📚 X_train_scaled Boyutu: (32950, 53)
🧪 X_test_scaled Boyutu : (8238, 53)

📊 ÖLÇEKLENDİRME ÖNCESİ VE SONRASI KARŞILAŞTIRMA:
----------------------------------------------------------------------
   İlk 5 özellik için istatistikler:

 Özellik  Önce Ort.  Önce Std.  Sonra Ort.  Sonra Std.
     age      40.06      10.43        -0.0         1.0
duration     258.03     257.16        -0.0         1.0
campaign       2.56       2.74         0.0         1.0
   pdays     962.20     187.57         0.0         1.0
previous       0.17       0.50         0.0         1.0
----------------------------------------------------------------------
   ✅ Tüm özellikler ortalama=0, std=1 olarak ölçeklendirildi!


### ✅ **Veri Hazırlık Özeti**

Veri hazırlama aşaması başarıyla tamamlandı! İşte yapılanlar:

#### **Tamamlanan İşlemler**
1. ✅ **Hedef Değişken Ayrıldı**: X (20 özellik) ve y ayrıldı
2. ✅ **Kategorik Encoding**: 10 kategorik değişken → One-Hot Encoding → 53 özelliğe çıkarıldı
3. ✅ **Train-Test Split**: %80 eğitim (32,950), %20 test (8,238) - Stratified
4. ✅ **Feature Scaling**: StandardScaler ile ortalama=0, std=1'e normalize edildi

#### **Hazır Veri Setleri**
- `X_train_scaled`: Eğitim özellikleri (32,950 × 53)
- `X_test_scaled`: Test özellikleri (8,238 × 53)
- `y_train`: Eğitim hedef değişkeni (32,950)
- `y_test`: Test hedef değişkeni (8,238)

#### **Önemli Notlar**
- 🔸 **Data Leakage Yok**: Scaling sadece eğitim setinden öğrenildi
- 🔸 **Dağılım Korundu**: Stratified split ile %88.7 / %11.3 oranı her iki sette de aynı
- 🔸 **Production Ready**: Scaler objesi model ile birlikte kaydedilmelidir

**Sırada:** Lojistik Regresyon modelinin matematiksel temellerini öğrenmek ve model eğitimi! 🚀

## 🤖 **Adım 6: Modeling - Lojistik Regresyon**

### 📚 **Lojistik Regresyon Nedir?**

**Lojistik Regresyon**, ikili (binary) sınıflandırma problemleri için kullanılan bir **denetimli öğrenme (supervised learning)** algoritmasıdır. Adında "regresyon" geçmesine rağmen aslında bir **sınıflandırma** algoritmasıdır.

**Kullanım Alanları:**
- Hastalık teşhisi (hasta/sağlıklı)
- Spam tespiti (spam/değil)
- Müşteri churn tahmini (ayrılır/ayrılmaz)
- Kredi riski değerlendirmesi (riskli/değil)
- **Bizim projemiz**: Müşteri term deposit alır mı? (evet/hayır)

### 🧮 **Matematiksel Temel**

#### **1. Lineer Regresyon vs Lojistik Regresyon**

**Lineer Regresyon:**
$$y = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + ... + \beta_n x_n$$

- Sonuç: $(-\infty, +\infty)$ aralığında herhangi bir değer
- Problem: Olasılık için uygun değil (0-1 arası olmalı)

**Lojistik Regresyon Çözümü:**
$$z = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + ... + \beta_n x_n$$

Bu lineer kombinasyonu **Sigmoid fonksiyonu** ile [0, 1] aralığına dönüştürürüz:

#### **2. Sigmoid Fonksiyonu**

$$\sigma(z) = \frac{1}{1 + e^{-z}}$$

**Özellikler:**
- Girdi: $(-\infty, +\infty)$
- Çıktı: $(0, 1)$ → Olasılık olarak yorumlanabilir
- $z = 0$ için $\sigma(z) = 0.5$
- $z \to +\infty$ için $\sigma(z) \to 1$
- $z \to -\infty$ için $\sigma(z) \to 0$

#### **3. Tahmin Formülü**

$$P(y=1|x) = \frac{1}{1 + e^{-(\beta_0 + \beta_1 x_1 + ... + \beta_n x_n)}}$$

Bu olasılık değerini bir sınıfa dönüştürmek için **threshold (eşik değeri)** kullanırız:

- $P(y=1|x) \geq 0.5$ → Sınıf = 1 (Pozitif)
- $P(y=1|x) < 0.5$ → Sınıf = 0 (Negatif)

#### **4. Cost Function (Maliyet Fonksiyonu)**

Lojistik regresyon **Log Loss (Binary Cross-Entropy)** kullanır:

$$J(\beta) = -\frac{1}{m} \sum_{i=1}^{m} [y_i \log(h(x_i)) + (1-y_i) \log(1-h(x_i))]$$

Burada:
- $m$: Örnek sayısı
- $y_i$: Gerçek etiket (0 veya 1)
- $h(x_i)$: Tahmin edilen olasılık

#### **5. Optimizasyon**

Cost function'ı minimize etmek için **Gradient Descent** veya **Newton-Raphson** gibi optimizasyon algoritmaları kullanılır.

### 📊 **Sigmoid Fonksiyonu Görselleştirme**

In [48]:
# Sigmoid fonksiyonu tanımla
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

# z değerleri oluştur
z_values = np.linspace(-10, 10, 200)
sigmoid_values = sigmoid(z_values)

# İnteraktif sigmoid grafiği
fig = go.Figure()

# Sigmoid eğrisi
fig.add_trace(go.Scatter(
    x=z_values,
    y=sigmoid_values,
    mode='lines',
    name='Sigmoid Function',
    line=dict(color='#FF69B4', width=3),
    hovertemplate='z: %{x:.2f}<br>σ(z): %{y:.3f}<extra></extra>'
))

# Threshold çizgisi (0.5)
fig.add_hline(y=0.5, line_dash="dash", line_color="red", 
              annotation_text="Threshold = 0.5", annotation_position="right")

# z=0 dikey çizgisi
fig.add_vline(x=0, line_dash="dash", line_color="gray",
              annotation_text="z = 0", annotation_position="top")

# Decision boundary bölgeleri
fig.add_hrect(y0=0.5, y1=1.0, fillcolor="#98FB98", opacity=0.2, 
              annotation_text="Sınıf 1 (Pozitif)", annotation_position="top right")
fig.add_hrect(y0=0.0, y1=0.5, fillcolor="#FFB6C1", opacity=0.2,
              annotation_text="Sınıf 0 (Negatif)", annotation_position="bottom right")

# Layout
fig.update_layout(
    title="<b>Sigmoid Fonksiyonu: σ(z) = 1 / (1 + e^(-z))</b>",
    title_font=dict(size=18, color='#2F4F4F'),
    xaxis_title="<b>z = β₀ + β₁x₁ + β₂x₂ + ... + βₙxₙ</b>",
    yaxis_title="<b>σ(z) = P(y=1|x)</b>",
    height=500,
    plot_bgcolor='white',
    paper_bgcolor='#F5F5F5',
    showlegend=True,
    hovermode='x unified'
)

fig.update_xaxes(showgrid=True, gridcolor='lightgray', zeroline=True)
fig.update_yaxes(showgrid=True, gridcolor='lightgray', range=[-0.05, 1.05])

fig.show()

print("✅ Sigmoid fonksiyonu görselleştirildi!")
print("\n💡 ÖNEMLİ NOKTALAR:")
print("   • z = 0 için σ(z) = 0.5 (exact)")
print("   • z > 0 için σ(z) > 0.5 → Sınıf 1")
print("   • z < 0 için σ(z) < 0.5 → Sınıf 0")
print("   • z → +∞ için σ(z) → 1")
print("   • z → -∞ için σ(z) → 0")

✅ Sigmoid fonksiyonu görselleştirildi!

💡 ÖNEMLİ NOKTALAR:
   • z = 0 için σ(z) = 0.5 (exact)
   • z > 0 için σ(z) > 0.5 → Sınıf 1
   • z < 0 için σ(z) < 0.5 → Sınıf 0
   • z → +∞ için σ(z) → 1
   • z → -∞ için σ(z) → 0


### 🎓 **Model Eğitimi (Training)**

Şimdi Scikit-learn kütüphanesini kullanarak Lojistik Regresyon modelimizi eğiteceğiz.

**Önemli Parametreler:**

- **C**: Regularization gücünün tersi (inverse). Düşük C = güçlü regularization
- **penalty**: L1 veya L2 regularization ('l1', 'l2', 'elasticnet', 'none')
- **solver**: Optimizasyon algoritması ('lbfgs', 'liblinear', 'saga', 'newton-cg')
- **max_iter**: Maksimum iterasyon sayısı
- **class_weight**: Class imbalance için ağırlık ('balanced' veya None)

**Class Imbalance Çözümü:** Veri setimizde %88.7 negatif, %11.3 pozitif var. `class_weight='balanced'` parametresi ile azınlık sınıfına daha fazla ağırlık vereceğiz.

In [50]:
from sklearn.linear_model import LogisticRegression
import time

print("🎓 LOJİSTİK REGRESYON MODELİ EĞİTİMİ BAŞLADI...")
print("="*70)

# Model oluştur
model = LogisticRegression(
    penalty='l2',              # L2 regularization
    C=1.0,                     # Regularization gücü (default)
    solver='lbfgs',            # Optimizasyon algoritması
    max_iter=1000,             # Maksimum iterasyon
    class_weight='balanced',   # Class imbalance için
    random_state=42            # Tekrar edilebilirlik
)

# Eğitim süresini ölç
start_time = time.time()

# Modeli eğit
model.fit(X_train_scaled, y_train)

end_time = time.time()
training_time = end_time - start_time

print(f"✅ MODEL EĞİTİMİ TAMAMLANDI!")
print("="*70)
print(f"⏱️  Eğitim Süresi              : {training_time:.3f} saniye")
print(f"🔢 Özellik Sayısı             : {model.n_features_in_}")
print(f"📊 Eğitim Veri Sayısı         : {X_train_scaled.shape[0]:,}")
print(f"🎯 Hedef Sınıf Sayısı         : {len(model.classes_)}")
print(f"🔄 Kullanılan İterasyon       : {model.n_iter_[0]}")
print("="*70)

# Eğitim seti üzerinde accuracy
train_accuracy = model.score(X_train_scaled, y_train)
print(f"\n📈 Eğitim Seti Doğruluğu      : {train_accuracy:.4f} ({train_accuracy*100:.2f}%)")

# Test seti üzerinde accuracy
test_accuracy = model.score(X_test_scaled, y_test)
print(f"🧪 Test Seti Doğruluğu         : {test_accuracy:.4f} ({test_accuracy*100:.2f}%)")
print("="*70)

🎓 LOJİSTİK REGRESYON MODELİ EĞİTİMİ BAŞLADI...
✅ MODEL EĞİTİMİ TAMAMLANDI!
⏱️  Eğitim Süresi              : 0.114 saniye
🔢 Özellik Sayısı             : 53
📊 Eğitim Veri Sayısı         : 32,950
🎯 Hedef Sınıf Sayısı         : 2
🔄 Kullanılan İterasyon       : 50

📈 Eğitim Seti Doğruluğu      : 0.8601 (86.01%)
🧪 Test Seti Doğruluğu         : 0.8626 (86.26%)


### 🔮 **Tahmin Yapma (Prediction)**

Lojistik regresyon modeli iki tip tahmin yapar:

1. **predict()**: Sınıf etiketi tahmin eder (0 veya 1)
2. **predict_proba()**: Her sınıf için olasılık tahmin eder [P(y=0), P(y=1)]

In [52]:
# Test seti üzerinde tahminler yap
y_pred = model.predict(X_test_scaled)
y_pred_proba = model.predict_proba(X_test_scaled)

print("🔮 TAHMİNLER OLUŞTURULDU")
print("="*70)
print(f"📊 Tahmin Sayısı: {len(y_pred):,}")
print("="*70)

# İlk 10 tahmin örneği
print("\n📋 İLK 10 TAHMİN ÖRNEĞİ:")
print("-"*70)
print(f"{'No':<5} {'Gerçek':>8} {'Tahmin':>8} {'P(y=0)':>10} {'P(y=1)':>10} {'Doğru?':>8}")
print("-"*70)

for i in range(10):
    correct = "✅" if y_test.iloc[i] == y_pred[i] else "❌"
    print(f"{i+1:<5} {y_test.iloc[i]:>8} {y_pred[i]:>8} {y_pred_proba[i][0]:>10.4f} {y_pred_proba[i][1]:>10.4f} {correct:>8}")

print("-"*70)

# Tahmin dağılımı
pred_counts = pd.Series(y_pred).value_counts()
print(f"\n📊 TAHMİN DAĞILIMI:")
print("-"*70)
print(f"   Sınıf 0 (Hayır): {pred_counts[0]:,} ({pred_counts[0]/len(y_pred)*100:.2f}%)")
print(f"   Sınıf 1 (Evet) : {pred_counts[1]:,} ({pred_counts[1]/len(y_pred)*100:.2f}%)")
print("="*70)

🔮 TAHMİNLER OLUŞTURULDU
📊 Tahmin Sayısı: 8,238

📋 İLK 10 TAHMİN ÖRNEĞİ:
----------------------------------------------------------------------
No      Gerçek   Tahmin     P(y=0)     P(y=1)   Doğru?
----------------------------------------------------------------------
1            0        0     0.9315     0.0685        ✅
2            0        0     0.9707     0.0293        ✅
3            0        0     0.9652     0.0348        ✅
4            0        1     0.0662     0.9338        ❌
5            0        0     0.6152     0.3848        ✅
6            0        1     0.2605     0.7395        ❌
7            0        0     0.9614     0.0386        ✅
8            0        0     0.7950     0.2050        ✅
9            1        1     0.0382     0.9618        ✅
10           0        0     0.9215     0.0785        ✅
----------------------------------------------------------------------

📊 TAHMİN DAĞILIMI:
----------------------------------------------------------------------
   Sınıf 0 (Hayır):

## 📊 **Adım 7: Model Evaluation - Model Değerlendirme**

Modelin performansını sadece accuracy ile değerlendirmek yetersizdir, özellikle dengesiz veri setlerinde. Aşağıdaki metrikleri kullanarak kapsamlı bir değerlendirme yapacağız:

### 🎯 **Değerlendirme Metrikleri**

#### **1. Confusion Matrix (Karmaşıklık Matrisi)**
- **True Positive (TP)**: Doğru tahmin edilen pozitif örnekler
- **True Negative (TN)**: Doğru tahmin edilen negatif örnekler
- **False Positive (FP)**: Yanlış pozitif (Type I Error)
- **False Negative (FN)**: Yanlış negatif (Type II Error)

#### **2. Classification Metrikleri**
- **Accuracy**: Genel doğruluk = (TP + TN) / (TP + TN + FP + FN)
- **Precision**: Kesinlik = TP / (TP + FP) - "Pozitif dediğimizin ne kadarı gerçekten pozitif?"
- **Recall (Sensitivity)**: Duyarlılık = TP / (TP + FN) - "Gerçek pozitiflerin ne kadarını bulduk?"
- **F1-Score**: Precision ve Recall'un harmonik ortalaması = 2 × (Precision × Recall) / (Precision + Recall)

#### **3. ROC-AUC**
- ROC Curve: True Positive Rate vs False Positive Rate grafiği
- AUC (Area Under Curve): 0.5-1.0 arası, yüksek = iyi model

### 📈 **Confusion Matrix ve Classification Report**

In [53]:
from sklearn.metrics import confusion_matrix, classification_report, ConfusionMatrixDisplay

# Confusion Matrix hesapla
cm = confusion_matrix(y_test, y_pred)

print("📊 CONFUSION MATRIX")
print("="*70)
print(cm)
print("="*70)

# Metrik değerlerini çıkar
tn, fp, fn, tp = cm.ravel()

print(f"\n📈 CONFUSION MATRIX BILEŞENLERI:")
print("-"*70)
print(f"   True Negative  (TN): {tn:>6,}  ✅ Doğru negatif tahminler")
print(f"   False Positive (FP): {fp:>6,}  ❌ Yanlış pozitif (Type I Error)")
print(f"   False Negative (FN): {fn:>6,}  ❌ Yanlış negatif (Type II Error)")
print(f"   True Positive  (TP): {tp:>6,}  ✅ Doğru pozitif tahminler")
print("-"*70)

# Manuel metrik hesaplama
accuracy = (tp + tn) / (tp + tn + fp + fn)
precision = tp / (tp + fp) if (tp + fp) > 0 else 0
recall = tp / (tp + fn) if (tp + fn) > 0 else 0
f1 = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0

print(f"\n📊 MANUEL HESAPLANAN METRİKLER:")
print("-"*70)
print(f"   Accuracy  : {accuracy:.4f} ({accuracy*100:.2f}%)")
print(f"   Precision : {precision:.4f} ({precision*100:.2f}%)")
print(f"   Recall    : {recall:.4f} ({recall*100:.2f}%)")
print(f"   F1-Score  : {f1:.4f} ({f1*100:.2f}%)")
print("="*70)

# Sklearn classification report
print(f"\n📋 CLASSIFICATION REPORT (SKLEARN):")
print("-"*70)
print(classification_report(y_test, y_pred, target_names=['Hayır (0)', 'Evet (1)']))

📊 CONFUSION MATRIX
[[6280 1030]
 [ 102  826]]

📈 CONFUSION MATRIX BILEŞENLERI:
----------------------------------------------------------------------
   True Negative  (TN):  6,280  ✅ Doğru negatif tahminler
   False Positive (FP):  1,030  ❌ Yanlış pozitif (Type I Error)
   False Negative (FN):    102  ❌ Yanlış negatif (Type II Error)
   True Positive  (TP):    826  ✅ Doğru pozitif tahminler
----------------------------------------------------------------------

📊 MANUEL HESAPLANAN METRİKLER:
----------------------------------------------------------------------
   Accuracy  : 0.8626 (86.26%)
   Precision : 0.4450 (44.50%)
   Recall    : 0.8901 (89.01%)
   F1-Score  : 0.5934 (59.34%)

📋 CLASSIFICATION REPORT (SKLEARN):
----------------------------------------------------------------------
              precision    recall  f1-score   support

   Hayır (0)       0.98      0.86      0.92      7310
    Evet (1)       0.45      0.89      0.59       928

    accuracy                        

In [ ]:
# İnteraktif Confusion Matrix Heatmap - Profesyonel Pastel Dizayn
# Pastel renk paleti: Yeşil tonları (başarı) ve Pembe tonları (hata)
pastel_colors = [
    [0.0, '#E8F5E9'],  # Çok açık yeşil
    [0.5, '#A5D6A7'],  # Açık yeşil  
    [1.0, '#66BB6A']   # Yeşil
]

fig = go.Figure(data=go.Heatmap(
    z=cm,
    x=['<b>Tahmin: Hayır (0)</b>', '<b>Tahmin: Evet (1)</b>'],
    y=['<b>Gerçek: Hayır (0)</b>', '<b>Gerçek: Evet (1)</b>'],
    colorscale=pastel_colors,
    showscale=True,
    colorbar=dict(
        title=dict(text="<b>Kayıt<br>Sayısı</b>", side="right"),
        thickness=20,
        len=0.7,
        tickfont=dict(size=11, family="Arial")
    ),
    hovertemplate='<b>%{y}</b><br><b>%{x}</b><br>Sayı: %{z:,}<extra></extra>',
    text=cm,
    texttemplate='<b>%{text:,}</b>',
    textfont=dict(size=18, family="Arial", color='#2C3E50')
))

# Profesyonel annotations ile metrik bilgileri
annotations = []

# TN (True Negative) - Sol üst
annotations.append(dict(
    x=0, y=0,
    text=f'<b>TN</b><br><span style="font-size:22px">{cm[0][0]:,}</span><br><span style="color:#27AE60">✓ Doğru</span>',
    showarrow=False,
    font=dict(size=13, family="Arial", color='#1B5E20'),
    bgcolor='rgba(232, 245, 233, 0.8)',
    bordercolor='#66BB6A',
    borderwidth=2,
    borderpad=8
))

# FP (False Positive) - Sağ üst
annotations.append(dict(
    x=1, y=0,
    text=f'<b>FP</b><br><span style="font-size:22px">{cm[0][1]:,}</span><br><span style="color:#E74C3C">✗ Type I Error</span>',
    showarrow=False,
    font=dict(size=13, family="Arial", color='#C62828'),
    bgcolor='rgba(255, 235, 238, 0.8)',
    bordercolor='#EF5350',
    borderwidth=2,
    borderpad=8
))

# FN (False Negative) - Sol alt
annotations.append(dict(
    x=0, y=1,
    text=f'<b>FN</b><br><span style="font-size:22px">{cm[1][0]:,}</span><br><span style="color:#E74C3C">✗ Type II Error</span>',
    showarrow=False,
    font=dict(size=13, family="Arial", color='#C62828'),
    bgcolor='rgba(255, 235, 238, 0.8)',
    bordercolor='#EF5350',
    borderwidth=2,
    borderpad=8
))

# TP (True Positive) - Sağ alt
annotations.append(dict(
    x=1, y=1,
    text=f'<b>TP</b><br><span style="font-size:22px">{cm[1][1]:,}</span><br><span style="color:#27AE60">✓ Doğru</span>',
    showarrow=False,
    font=dict(size=13, family="Arial", color='#1B5E20'),
    bgcolor='rgba(232, 245, 233, 0.8)',
    bordercolor='#66BB6A',
    borderwidth=2,
    borderpad=8
))

fig.update_layout(
    title=dict(
        text="<b>Confusion Matrix - Model Performans Analizi</b>",
        font=dict(size=20, family="Arial", color='#2C3E50'),
        x=0.5,
        xanchor='center'
    ),
    xaxis=dict(
        title=dict(
            text="<b>Tahmin Edilen Sınıf</b>",
            font=dict(size=14, family="Arial", color='#34495E')
        ),
        tickfont=dict(size=12, family="Arial"),
        side='bottom'
    ),
    yaxis=dict(
        title=dict(
            text="<b>Gerçek Sınıf</b>",
            font=dict(size=14, family="Arial", color='#34495E')
        ),
        tickfont=dict(size=12, family="Arial")
    ),
    plot_bgcolor='#FAFAFA',
    paper_bgcolor='#FFFFFF',
    annotations=annotations,
    margin=dict(l=100, r=100, t=80, b=80)
)

fig.update_xaxes(showgrid=False)
fig.update_yaxes(showgrid=False)

fig.show()

print("\n✅ Profesyonel Confusion Matrix görselleştirildi!")
print("\n📊 Confusion Matrix Bileşenleri:")
print(f"   • TN (True Negative)  : {cm[0][0]:,} - Doğru negatif tahminler")
print(f"   • FP (False Positive) : {cm[0][1]:,} - Yanlış pozitif (Type I Error)")
print(f"   • FN (False Negative) : {cm[1][0]:,} - Yanlış negatif (Type II Error)")
print(f"   • TP (True Positive)  : {cm[1][1]:,} - Doğru pozitif tahminler")


✅ Profesyonel Confusion Matrix görselleştirildi!

📊 Confusion Matrix Bileşenleri:
   • TN (True Negative)  : 6,280 - Doğru negatif tahminler
   • FP (False Positive) : 1,030 - Yanlış pozitif (Type I Error)
   • FN (False Negative) : 102 - Yanlış negatif (Type II Error)
   • TP (True Positive)  : 826 - Doğru pozitif tahminler


### 📉 **ROC Curve ve AUC Score**

**ROC (Receiver Operating Characteristic) Curve**, farklı threshold değerleri için True Positive Rate (TPR) ve False Positive Rate (FPR) değerlerini gösteren bir grafiktir.

- **TPR (Sensitivity/Recall)**: TP / (TP + FN) - Gerçek pozitifleri bulma oranı
- **FPR**: FP / (FP + TN) - Yanlış pozitif oranı
- **AUC (Area Under Curve)**: 0.5-1.0 arası, 1.0'a yakın = mükemmel model

In [58]:
from sklearn.metrics import roc_curve, roc_auc_score

# ROC Curve için FPR, TPR hesapla
y_pred_proba_pos = y_pred_proba[:, 1]  # Pozitif sınıf olasılıkları
fpr, tpr, thresholds = roc_curve(y_test, y_pred_proba_pos)

# AUC skorunu hesapla
auc_score = roc_auc_score(y_test, y_pred_proba_pos)

print("📉 ROC-AUC ANALİZİ")
print("="*70)
print(f"📊 AUC Score: {auc_score:.4f}")
print("="*70)

# ROC Curve Görselleştirme
fig = go.Figure()

# ROC Curve
fig.add_trace(go.Scatter(
    x=fpr,
    y=tpr,
    mode='lines',
    name=f'ROC Curve (AUC = {auc_score:.4f})',
    line=dict(color='#FF69B4', width=3),
    hovertemplate='FPR: %{x:.3f}<br>TPR: %{y:.3f}<extra></extra>'
))

# Rastgele tahmin çizgisi (diagonal)
fig.add_trace(go.Scatter(
    x=[0, 1],
    y=[0, 1],
    mode='lines',
    name='Rastgele Tahmin (AUC = 0.50)',
    line=dict(color='gray', width=2, dash='dash'),
    hovertemplate='Rastgele<extra></extra>'
))

# AUC alanını vurgula
fig.add_trace(go.Scatter(
    x=fpr,
    y=tpr,
    fill='tozeroy',
    fillcolor='rgba(255, 182, 193, 0.3)',
    line=dict(width=0),
    showlegend=False,
    hoverinfo='skip'
))

# Layout
fig.update_layout(
    title=f"<b>ROC Curve - AUC Score: {auc_score:.4f}</b>",
    title_font=dict(size=18, color='#2F4F4F'),
    xaxis_title="<b>False Positive Rate (FPR)</b>",
    yaxis_title="<b>True Positive Rate (TPR) / Recall</b>",
    height=600,
    width=700,
    plot_bgcolor='white',
    paper_bgcolor='#F5F5F5',
    showlegend=True,
    legend=dict(x=0.6, y=0.1),
    hovermode='closest'
)

fig.update_xaxes(showgrid=True, gridcolor='lightgray', range=[-0.02, 1.02])
fig.update_yaxes(showgrid=True, gridcolor='lightgray', range=[-0.02, 1.02])

fig.show()

print("\n✅ ROC Curve görselleştirildi!")
print("\n💡 AUC SCORE YORUMLAMA:")
print("-"*70)
print("   • 0.90 - 1.00: Mükemmel")
print("   • 0.80 - 0.90: İyi")
print("   • 0.70 - 0.80: Orta")
print("   • 0.60 - 0.70: Zayıf")
print("   • 0.50 - 0.60: Başarısız")
print("-"*70)
print(f"   Modelimizin AUC'si: {auc_score:.4f} → İYİ SEVİYEDE!")
print("="*70)

📉 ROC-AUC ANALİZİ
📊 AUC Score: 0.9396



✅ ROC Curve görselleştirildi!

💡 AUC SCORE YORUMLAMA:
----------------------------------------------------------------------
   • 0.90 - 1.00: Mükemmel
   • 0.80 - 0.90: İyi
   • 0.70 - 0.80: Orta
   • 0.60 - 0.70: Zayıf
   • 0.50 - 0.60: Başarısız
----------------------------------------------------------------------
   Modelimizin AUC'si: 0.9396 → İYİ SEVİYEDE!


### 🔑 **Feature Importance (Özellik Önemleri)**

Lojistik regresyonda **katsayılar (coefficients)** özelliklerin önemini gösterir:

- **Pozitif katsayı**: Özellik arttıkça pozitif sınıf olasılığı artar
- **Negatif katsayı**: Özellik arttıkça negatif sınıf olasılığı artar
- **Büyük mutlak değer**: Özellik daha etkili

In [60]:
# Katsayıları çıkar
coefficients = model.coef_[0]
feature_names = X_encoded.columns

# DataFrame oluştur
feature_importance_df = pd.DataFrame({
    'Özellik': feature_names,
    'Katsayı': coefficients,
    'Mutlak Değer': np.abs(coefficients)
}).sort_values('Mutlak Değer', ascending=False)

# En önemli 15 özellik
top_15 = feature_importance_df.head(15)

print("🔑 FEATURE IMPORTANCE ANALİZİ")
print("="*70)
print(f"📊 Toplam Özellik Sayısı: {len(feature_importance_df)}")
print("="*70)

print("\n📈 EN ÖNEMLİ 15 ÖZELLİK:")
print("-"*70)
print(f"{'Sıra':<5} {'Özellik':<35} {'Katsayı':>12} {'Etki':>10}")
print("-"*70)

for idx, row in top_15.iterrows():
    effect = "POZITIF ↑" if row['Katsayı'] > 0 else "NEGATIF ↓"
    print(f"{feature_importance_df.index.tolist().index(idx)+1:<5} {row['Özellik']:<35} {row['Katsayı']:>12.4f} {effect:>10}")

print("="*70)

# Görselleştirme: En önemli 15 özellik
fig = go.Figure()

colors = ['#98FB98' if c > 0 else '#FFB6C1' for c in top_15['Katsayı']]

fig.add_trace(go.Bar(
    y=top_15['Özellik'][::-1],  # Ters sıralama (en büyük üstte)
    x=top_15['Katsayı'][::-1],
    orientation='h',
    marker=dict(color=colors[::-1]),
    text=top_15['Katsayı'][::-1].round(3),
    textposition='outside',
    hovertemplate='%{y}<br>Katsayı: %{x:.4f}<extra></extra>'
))

fig.update_layout(
    title="<b>En Önemli 15 Özellik (Katsayılarına Göre)</b>",
    title_font=dict(size=18, color='#2F4F4F'),
    xaxis_title="<b>Katsayı Değeri</b>",
    yaxis_title="<b>Özellik Adı</b>",
    height=700,
    plot_bgcolor='white',
    paper_bgcolor='#F5F5F5',
    showlegend=False
)

fig.add_vline(x=0, line_width=2, line_color="black")
fig.update_xaxes(showgrid=True, gridcolor='lightgray', zeroline=True)
fig.update_yaxes(showgrid=False)

fig.show()

print("\n✅ Feature importance görselleştirildi!")
print("\n💡 YORUMLAMA:")
print("-"*70)
print("   🟢 Yeşil (Pozitif): Özellik arttıkça 'Evet' olasılığı artar")
print("   🔴 Pembe (Negatif): Özellik arttıkça 'Hayır' olasılığı artar")
print("="*70)

🔑 FEATURE IMPORTANCE ANALİZİ
📊 Toplam Özellik Sayısı: 53

📈 EN ÖNEMLİ 15 ÖZELLİK:
----------------------------------------------------------------------
Sıra  Özellik                                  Katsayı       Etki
----------------------------------------------------------------------
1     emp_var_rate                             -3.5989  NEGATIF ↓
2     duration                                  1.8028  POZITIF ↑
3     cons_price_idx                            1.3801  POZITIF ↑
4     euribor3m                                 1.1655  POZITIF ↑
5     month_may                                -0.3855  NEGATIF ↓
6     month_aug                                 0.3799  POZITIF ↑
7     month_jun                                -0.3034  NEGATIF ↓
8     nr_employed                               0.2790  POZITIF ↑
9     month_mar                                 0.2456  POZITIF ↑
10    poutcome_success                          0.2215  POZITIF ↑
11    contact_telephone                        -0.


✅ Feature importance görselleştirildi!

💡 YORUMLAMA:
----------------------------------------------------------------------
   🟢 Yeşil (Pozitif): Özellik arttıkça 'Evet' olasılığı artar
   🔴 Pembe (Negatif): Özellik arttıkça 'Hayır' olasılığı artar


## 💾 **Adım 8: Model Kaydetme (Model Saving)**

Eğitilmiş modeli production ortamında kullanmak için kaydetmeliyiz. Python'da iki popüler yöntem vardır:

### **1. Pickle (Python Standart Kütüphanesi)**
- Python'un built-in modülü
- Herhangi bir Python objesini serialize edebilir
- Dosya uzantısı: `.pkl` veya `.pickle`

### **2. Joblib (Scikit-learn Tavsiyesi)**
- Büyük NumPy array'leri için daha verimli
- Scikit-learn modelleri için özellikle optimize edilmiş
- Dosya uzantısı: `.joblib` veya `.pkl`

**Production için gerekli objeler:**
- ✅ Eğitilmiş model (LogisticRegression)
- ✅ Scaler objesi (StandardScaler)
- ✅ Feature names (sütun isimleri)

In [62]:
import pickle
import joblib
import os

# Model klasörü oluştur
model_dir = 'models'
os.makedirs(model_dir, exist_ok=True)

print("💾 MODEL KAYDETME İŞLEMİ BAŞLADI...")
print("="*70)

# 1. Pickle ile kaydetme
pickle_model_path = os.path.join(model_dir, 'logistic_regression_model.pkl')
pickle_scaler_path = os.path.join(model_dir, 'scaler.pkl')

with open(pickle_model_path, 'wb') as f:
    pickle.dump(model, f)
    
with open(pickle_scaler_path, 'wb') as f:
    pickle.dump(scaler, f)

print(f"✅ Pickle ile kaydedildi:")
print(f"   Model  : {pickle_model_path}")
print(f"   Scaler : {pickle_scaler_path}")
print()

# 2. Joblib ile kaydetme
joblib_model_path = os.path.join(model_dir, 'logistic_regression_model.joblib')
joblib_scaler_path = os.path.join(model_dir, 'scaler.joblib')

joblib.dump(model, joblib_model_path)
joblib.dump(scaler, joblib_scaler_path)

print(f"✅ Joblib ile kaydedildi:")
print(f"   Model  : {joblib_model_path}")
print(f"   Scaler : {joblib_scaler_path}")
print()

# 3. Feature names kaydetme
feature_names_path = os.path.join(model_dir, 'feature_names.pkl')
with open(feature_names_path, 'wb') as f:
    pickle.dump(list(X_encoded.columns), f)

print(f"✅ Feature names kaydedildi:")
print(f"   Features: {feature_names_path}")
print("="*70)

# Dosya boyutlarını kontrol et
pickle_size = os.path.getsize(pickle_model_path) / 1024  # KB
joblib_size = os.path.getsize(joblib_model_path) / 1024  # KB

print(f"\n📊 DOSYA BOYUTLARI:")
print("-"*70)
print(f"   Pickle Model : {pickle_size:.2f} KB")
print(f"   Joblib Model : {joblib_size:.2f} KB")
print("-"*70)
print("="*70)

💾 MODEL KAYDETME İŞLEMİ BAŞLADI...
✅ Pickle ile kaydedildi:
   Model  : models\logistic_regression_model.pkl
   Scaler : models\scaler.pkl

✅ Joblib ile kaydedildi:
   Model  : models\logistic_regression_model.joblib
   Scaler : models\scaler.joblib

✅ Feature names kaydedildi:
   Features: models\feature_names.pkl

📊 DOSYA BOYUTLARI:
----------------------------------------------------------------------
   Pickle Model : 1.13 KB
   Joblib Model : 1.28 KB
----------------------------------------------------------------------


### 🔄 **Model Yükleme ve Test Etme**

Kaydedilen modeli yükleyip çalıştığını doğrulayalım.

In [64]:
# Joblib ile yükleme (önerilen)
print("🔄 MODEL YÜKLEME İŞLEMİ...")
print("="*70)

loaded_model = joblib.load(joblib_model_path)
loaded_scaler = joblib.load(joblib_scaler_path)

with open(feature_names_path, 'rb') as f:
    loaded_feature_names = pickle.load(f)

print("✅ Model ve scaler başarıyla yüklendi!")
print("="*70)

# Test: Yüklenen model ile tahmin yap
test_sample = X_test_scaled[:5]  # İlk 5 örnek
test_predictions = loaded_model.predict(test_sample)
test_probabilities = loaded_model.predict_proba(test_sample)

print("\n🧪 YÜKLENEN MODEL TEST EDİLİYOR...")
print("-"*70)
print(f"Test örnek sayısı: {len(test_sample)}")
print(f"Tahminler: {test_predictions}")
print(f"\nOlasılıklar:")
for i, (pred, probs) in enumerate(zip(test_predictions, test_probabilities), 1):
    print(f"   Örnek {i}: Sınıf {pred} - P(0)={probs[0]:.4f}, P(1)={probs[1]:.4f}")

print("-"*70)

# Doğruluk kontrolü: Orijinal model ile aynı sonuçları üretmeli
original_pred = model.predict(test_sample)
is_same = np.array_equal(test_predictions, original_pred)

print(f"\n🔍 DOĞRULAMA:")
print(f"   Yüklenen model ile orijinal model aynı sonuçları verdi mi? {' YES ✅' if is_same else ' NO ❌'}")
print("="*70)

print("\n💡 MODEL KULLANIMI İÇİN PYTHON KODU:")
print("-"*70)
print("""
# 1. Gerekli kütüphaneleri import et
import joblib
import pandas as pd
import numpy as np

# 2. Modeli ve scaler'ı yükle
model = joblib.load('models/logistic_regression_model.joblib')
scaler = joblib.load('models/scaler.joblib')

# 3. Yeni veri geldiğinde:
# - Önce aynı preprocessing (encoding) uygula
# - Scaler ile ölçeklendir
# - Model ile tahmin yap

new_data_scaled = scaler.transform(new_data_encoded)
prediction = model.predict(new_data_scaled)
probability = model.predict_proba(new_data_scaled)
""")

🔄 MODEL YÜKLEME İŞLEMİ...
✅ Model ve scaler başarıyla yüklendi!

🧪 YÜKLENEN MODEL TEST EDİLİYOR...
----------------------------------------------------------------------
Test örnek sayısı: 5
Tahminler: [0 0 0 1 0]

Olasılıklar:
   Örnek 1: Sınıf 0 - P(0)=0.9315, P(1)=0.0685
   Örnek 2: Sınıf 0 - P(0)=0.9707, P(1)=0.0293
   Örnek 3: Sınıf 0 - P(0)=0.9652, P(1)=0.0348
   Örnek 4: Sınıf 1 - P(0)=0.0662, P(1)=0.9338
   Örnek 5: Sınıf 0 - P(0)=0.6152, P(1)=0.3848
----------------------------------------------------------------------

🔍 DOĞRULAMA:
   Yüklenen model ile orijinal model aynı sonuçları verdi mi?  YES ✅

💡 MODEL KULLANIMI İÇİN PYTHON KODU:
----------------------------------------------------------------------

# 1. Gerekli kütüphaneleri import et
import joblib
import pandas as pd
import numpy as np

# 2. Modeli ve scaler'ı yükle
model = joblib.load('models/logistic_regression_model.joblib')
scaler = joblib.load('models/scaler.joblib')

# 3. Yeni veri geldiğinde:
# - Önce aynı pre